// @ariakit/react-core/cjs/__chunks/2WCJJQSK.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});"use client";



var _UZNYSPKPcjs = require('./UZNYSPKP.cjs');



var _UVBBMANLcjs = require('./UVBBMANL.cjs');


var _RNZNGEL4cjs = require('./RNZNGEL4.cjs');

// src/select/select-context.tsx
var _react = require('react');
var ctx = _RNZNGEL4cjs.createStoreContext.call(void 0, 
  [_UVBBMANLcjs.PopoverContextProvider, _UZNYSPKPcjs.CompositeContextProvider],
  [_UVBBMANLcjs.PopoverScopedContextProvider, _UZNYSPKPcjs.CompositeScopedContextProvider]
);
var useSelectContext = ctx.useContext;
var useSelectScopedContext = ctx.useScopedContext;
var useSelectProviderContext = ctx.useProviderContext;
var SelectContextProvider = ctx.ContextProvider;
var SelectScopedContextProvider = ctx.ScopedContextProvider;
var SelectItemCheckedContext = _react.createContext.call(void 0, false);








exports.useSelectContext = useSelectContext; exports.useSelectScopedContext = useSelectScopedContext; exports.useSelectProviderContext = useSelectProviderContext; exports.SelectContextProvider = SelectContextProvider; exports.SelectScopedContextProvider = SelectScopedContextProvider; exports.SelectItemCheckedContext = SelectItemCheckedContext;
