// @ariakit/react-core/cjs/__chunks/3WCBE6SU.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});"use client";



var _UZNYSPKPcjs = require('./UZNYSPKP.cjs');



var _UVBBMANLcjs = require('./UVBBMANL.cjs');


var _RNZNGEL4cjs = require('./RNZNGEL4.cjs');

// src/combobox/combobox-context.tsx
var _react = require('react');
var ctx = _RNZNGEL4cjs.createStoreContext.call(void 0, 
  [_UVBBMANLcjs.PopoverContextProvider, _UZNYSPKPcjs.CompositeContextProvider],
  [_UVBBMANLcjs.PopoverScopedContextProvider, _UZNYSPKPcjs.CompositeScopedContextProvider]
);
var useComboboxContext = ctx.useContext;
var useComboboxScopedContext = ctx.useScopedContext;
var useComboboxProviderContext = ctx.useProviderContext;
var ComboboxContextProvider = ctx.ContextProvider;
var ComboboxScopedContextProvider = ctx.ScopedContextProvider;
var ComboboxItemValueContext = _react.createContext.call(void 0, 
  void 0
);
var ComboboxItemCheckedContext = _react.createContext.call(void 0, false);









exports.useComboboxContext = useComboboxContext; exports.useComboboxScopedContext = useComboboxScopedContext; exports.useComboboxProviderContext = useComboboxProviderContext; exports.ComboboxContextProvider = ComboboxContextProvider; exports.ComboboxScopedContextProvider = ComboboxScopedContextProvider; exports.ComboboxItemValueContext = ComboboxItemValueContext; exports.ComboboxItemCheckedContext = ComboboxItemCheckedContext;
