// @ariakit/react-core/cjs/__chunks/7LBZXSLY.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});"use client";



var _UZNYSPKPcjs = require('./UZNYSPKP.cjs');


var _RNZNGEL4cjs = require('./RNZNGEL4.cjs');

// src/menubar/menubar-context.ts
var _react = require('react');
var menubar = _RNZNGEL4cjs.createStoreContext.call(void 0, 
  [_UZNYSPKPcjs.CompositeContextProvider],
  [_UZNYSPKPcjs.CompositeScopedContextProvider]
);
var useMenubarContext = menubar.useContext;
var useMenubarScopedContext = menubar.useScopedContext;
var useMenubarProviderContext = menubar.useProviderContext;
var MenubarContextProvider = menubar.ContextProvider;
var MenubarScopedContextProvider = menubar.ScopedContextProvider;
var MenuItemCheckedContext = _react.createContext.call(void 0, 
  void 0
);








exports.useMenubarContext = useMenubarContext; exports.useMenubarScopedContext = useMenubarScopedContext; exports.useMenubarProviderContext = useMenubarProviderContext; exports.MenubarContextProvider = MenubarContextProvider; exports.MenubarScopedContextProvider = MenubarScopedContextProvider; exports.MenuItemCheckedContext = MenuItemCheckedContext;
