// @ariakit/react-core/cjs/__chunks/F2A2ZQDB.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});"use client";



var _S6UU7NA4cjs = require('./S6UU7NA4.cjs');


var _RNZNGEL4cjs = require('./RNZNGEL4.cjs');

// src/dialog/dialog-context.tsx
var _react = require('react');
var ctx = _RNZNGEL4cjs.createStoreContext.call(void 0, 
  [_S6UU7NA4cjs.DisclosureContextProvider],
  [_S6UU7NA4cjs.DisclosureScopedContextProvider]
);
var useDialogContext = ctx.useContext;
var useDialogScopedContext = ctx.useScopedContext;
var useDialogProviderContext = ctx.useProviderContext;
var DialogContextProvider = ctx.ContextProvider;
var DialogScopedContextProvider = ctx.ScopedContextProvider;
var DialogHeadingContext = _react.createContext.call(void 0, void 0);
var DialogDescriptionContext = _react.createContext.call(void 0, void 0);









exports.useDialogContext = useDialogContext; exports.useDialogScopedContext = useDialogScopedContext; exports.useDialogProviderContext = useDialogProviderContext; exports.DialogContextProvider = DialogContextProvider; exports.DialogScopedContextProvider = DialogScopedContextProvider; exports.DialogHeadingContext = DialogHeadingContext; exports.DialogDescriptionContext = DialogDescriptionContext;
