// @ariakit/react-core/cjs/__chunks/NGEKJJDH.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});"use client";






var _7LBZXSLYcjs = require('./7LBZXSLY.cjs');



var _UZNYSPKPcjs = require('./UZNYSPKP.cjs');



var _3W6FH6JKcjs = require('./3W6FH6JK.cjs');


var _RNZNGEL4cjs = require('./RNZNGEL4.cjs');

// src/menu/menu-context.ts
var _react = require('react');
var menu = _RNZNGEL4cjs.createStoreContext.call(void 0, 
  [_UZNYSPKPcjs.CompositeContextProvider, _3W6FH6JKcjs.HovercardContextProvider],
  [_UZNYSPKPcjs.CompositeScopedContextProvider, _3W6FH6JKcjs.HovercardScopedContextProvider]
);
var useMenuContext = menu.useContext;
var useMenuScopedContext = menu.useScopedContext;
var useMenuProviderContext = menu.useProviderContext;
var MenuContextProvider = menu.ContextProvider;
var MenuScopedContextProvider = menu.ScopedContextProvider;
var useMenuBarContext = _7LBZXSLYcjs.useMenubarContext;
var useMenuBarScopedContext = _7LBZXSLYcjs.useMenubarScopedContext;
var useMenuBarProviderContext = _7LBZXSLYcjs.useMenubarProviderContext;
var MenuBarContextProvider = _7LBZXSLYcjs.MenubarContextProvider;
var MenuBarScopedContextProvider = _7LBZXSLYcjs.MenubarScopedContextProvider;
var MenuItemCheckedContext = _react.createContext.call(void 0, 
  void 0
);













exports.useMenuContext = useMenuContext; exports.useMenuScopedContext = useMenuScopedContext; exports.useMenuProviderContext = useMenuProviderContext; exports.MenuContextProvider = MenuContextProvider; exports.MenuScopedContextProvider = MenuScopedContextProvider; exports.useMenuBarContext = useMenuBarContext; exports.useMenuBarScopedContext = useMenuBarScopedContext; exports.useMenuBarProviderContext = useMenuBarProviderContext; exports.MenuBarContextProvider = MenuBarContextProvider; exports.MenuBarScopedContextProvider = MenuBarScopedContextProvider; exports.MenuItemCheckedContext = MenuItemCheckedContext;
