// @ariakit/react-core/cjs/__chunks/R66IWXK6.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});"use client";




var _RNZNGEL4cjs = require('./RNZNGEL4.cjs');



var _AV6KTKLEcjs = require('./AV6KTKLE.cjs');

// src/visually-hidden/visually-hidden.ts
var useVisuallyHidden = _RNZNGEL4cjs.createHook.call(void 0, (props) => {
  props = _AV6KTKLEcjs.__spreadProps.call(void 0, _AV6KTKLEcjs.__spreadValues.call(void 0, {}, props), {
    style: _AV6KTKLEcjs.__spreadValues.call(void 0, {
      border: 0,
      clip: "rect(0 0 0 0)",
      height: "1px",
      margin: "-1px",
      overflow: "hidden",
      padding: 0,
      position: "absolute",
      whiteSpace: "nowrap",
      width: "1px"
    }, props.style)
  });
  return props;
});
var VisuallyHidden = _RNZNGEL4cjs.createComponent.call(void 0, 
  (props) => {
    const htmlProps = useVisuallyHidden(props);
    return _RNZNGEL4cjs.createElement.call(void 0, "span", htmlProps);
  }
);
if (process.env.NODE_ENV !== "production") {
  VisuallyHidden.displayName = "VisuallyHidden";
}




exports.useVisuallyHidden = useVisuallyHidden; exports.VisuallyHidden = VisuallyHidden;
