// @ariakit/react-core/cjs/__chunks/T4KYZJUY.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});"use client";


var _ERFHNHONcjs = require('./ERFHNHON.cjs');




var _RNZNGEL4cjs = require('./RNZNGEL4.cjs');




var _AV6KTKLEcjs = require('./AV6KTKLE.cjs');

// src/checkbox/checkbox-check.tsx
var _react = require('react');
var _jsxruntime = require('react/jsx-runtime');
var checkmark = /* @__PURE__ */ _jsxruntime.jsx.call(void 0, 
  "svg",
  {
    display: "block",
    fill: "none",
    stroke: "currentColor",
    strokeLinecap: "round",
    strokeLinejoin: "round",
    strokeWidth: "1.5pt",
    viewBox: "0 0 16 16",
    height: "1em",
    width: "1em",
    children: /* @__PURE__ */ _jsxruntime.jsx.call(void 0, "polyline", { points: "4,8 7,12 12,4" })
  }
);
function getChildren(props) {
  if (props.checked) {
    return props.children || checkmark;
  }
  if (typeof props.children === "function") {
    return props.children;
  }
  return null;
}
var useCheckboxCheck = _RNZNGEL4cjs.createHook.call(void 0, 
  (_a) => {
    var _b = _a, { store, checked } = _b, props = _AV6KTKLEcjs.__objRest.call(void 0, _b, ["store", "checked"]);
    const context = _react.useContext.call(void 0, _ERFHNHONcjs.CheckboxCheckedContext);
    checked = checked != null ? checked : context;
    const children = getChildren({ checked, children: props.children });
    props = _AV6KTKLEcjs.__spreadProps.call(void 0, _AV6KTKLEcjs.__spreadValues.call(void 0, {
      "aria-hidden": true
    }, props), {
      children,
      style: _AV6KTKLEcjs.__spreadValues.call(void 0, {
        width: "1em",
        height: "1em",
        pointerEvents: "none"
      }, props.style)
    });
    return props;
  }
);
var CheckboxCheck = _RNZNGEL4cjs.createComponent.call(void 0, (props) => {
  const htmlProps = useCheckboxCheck(props);
  return _RNZNGEL4cjs.createElement.call(void 0, "span", htmlProps);
});
if (process.env.NODE_ENV !== "production") {
  CheckboxCheck.displayName = "CheckboxCheck";
}




exports.useCheckboxCheck = useCheckboxCheck; exports.CheckboxCheck = CheckboxCheck;
