// @ariakit/react-core/cjs/__chunks/WZ3DRKGP.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});"use client";


var _VV6WA3I6cjs = require('./VV6WA3I6.cjs');


var _LAUATD5Ocjs = require('./LAUATD5O.cjs');


var _YPVQYY4Jcjs = require('./YPVQYY4J.cjs');


var _JAQJG42Rcjs = require('./JAQJG42R.cjs');



var _EO6LS72Hcjs = require('./EO6LS72H.cjs');



var _AV6KTKLEcjs = require('./AV6KTKLE.cjs');

// src/dialog/dialog-backdrop.tsx
var _react = require('react');
var _jsxruntime = require('react/jsx-runtime');
function DialogBackdrop({
  store,
  backdrop,
  backdropProps,
  alwaysVisible,
  hidden
}) {
  const ref = _react.useRef.call(void 0, null);
  const disclosure = _JAQJG42Rcjs.useDisclosureStore.call(void 0, { disclosure: store });
  const contentElement = store.useState("contentElement");
  _EO6LS72Hcjs.useSafeLayoutEffect.call(void 0, () => {
    const backdrop2 = ref.current;
    const dialog = contentElement;
    if (!backdrop2)
      return;
    if (!dialog)
      return;
    backdrop2.style.zIndex = getComputedStyle(dialog).zIndex;
  }, [contentElement]);
  _EO6LS72Hcjs.useSafeLayoutEffect.call(void 0, () => {
    const id = contentElement == null ? void 0 : contentElement.id;
    if (!id)
      return;
    const backdrop2 = ref.current;
    if (!backdrop2)
      return;
    return _VV6WA3I6cjs.markAncestor.call(void 0, backdrop2, id);
  }, [contentElement]);
  if (hidden != null) {
    backdropProps = _AV6KTKLEcjs.__spreadProps.call(void 0, _AV6KTKLEcjs.__spreadValues.call(void 0, {}, backdropProps), { hidden });
  }
  const props = _YPVQYY4Jcjs.useDisclosureContent.call(void 0, _AV6KTKLEcjs.__spreadProps.call(void 0, _AV6KTKLEcjs.__spreadValues.call(void 0, {
    store: disclosure,
    role: "presentation",
    "data-backdrop": (contentElement == null ? void 0 : contentElement.id) || "",
    alwaysVisible
  }, backdropProps), {
    ref: _EO6LS72Hcjs.useMergeRefs.call(void 0, backdropProps == null ? void 0 : backdropProps.ref, ref),
    style: _AV6KTKLEcjs.__spreadValues.call(void 0, {
      position: "fixed",
      top: 0,
      right: 0,
      bottom: 0,
      left: 0
    }, backdropProps == null ? void 0 : backdropProps.style)
  }));
  if (!backdrop)
    return null;
  if (_react.isValidElement.call(void 0, backdrop)) {
    return /* @__PURE__ */ _jsxruntime.jsx.call(void 0, _LAUATD5Ocjs.Role, _AV6KTKLEcjs.__spreadProps.call(void 0, _AV6KTKLEcjs.__spreadValues.call(void 0, {}, props), { render: backdrop }));
  }
  const Component = typeof backdrop !== "boolean" ? backdrop : "div";
  return /* @__PURE__ */ _jsxruntime.jsx.call(void 0, _LAUATD5Ocjs.Role, _AV6KTKLEcjs.__spreadProps.call(void 0, _AV6KTKLEcjs.__spreadValues.call(void 0, {}, props), { render: /* @__PURE__ */ _jsxruntime.jsx.call(void 0, Component, {}) }));
}



exports.DialogBackdrop = DialogBackdrop;
