// @ariakit/react-core/cjs/__chunks/XB3G2EO2.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});"use client";


var _R66IWXK6cjs = require('./R66IWXK6.cjs');




var _RNZNGEL4cjs = require('./RNZNGEL4.cjs');



var _AV6KTKLEcjs = require('./AV6KTKLE.cjs');

// src/focus-trap/focus-trap.ts
var useFocusTrap = _RNZNGEL4cjs.createHook.call(void 0, (props) => {
  props = _AV6KTKLEcjs.__spreadProps.call(void 0, _AV6KTKLEcjs.__spreadValues.call(void 0, {
    "data-focus-trap": "",
    tabIndex: 0,
    "aria-hidden": true
  }, props), {
    style: _AV6KTKLEcjs.__spreadValues.call(void 0, {
      // Prevents unintended scroll jumps.
      position: "fixed",
      top: 0,
      left: 0
    }, props.style)
  });
  props = _R66IWXK6cjs.useVisuallyHidden.call(void 0, props);
  return props;
});
var FocusTrap = _RNZNGEL4cjs.createComponent.call(void 0, (props) => {
  const htmlProps = useFocusTrap(props);
  return _RNZNGEL4cjs.createElement.call(void 0, "span", htmlProps);
});
if (process.env.NODE_ENV !== "production") {
  FocusTrap.displayName = "FocusTrap";
}




exports.useFocusTrap = useFocusTrap; exports.FocusTrap = FocusTrap;
