// @ariakit/react-core/cjs/__chunks/YU7XEYUJ.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});"use client";


var _UVBBMANLcjs = require('./UVBBMANL.cjs');




var _RNZNGEL4cjs = require('./RNZNGEL4.cjs');




var _AV6KTKLEcjs = require('./AV6KTKLE.cjs');

// src/popover/popover-disclosure-arrow.tsx
var _react = require('react');
var _misc = require('@ariakit/core/utils/misc');
var _jsxruntime = require('react/jsx-runtime');
var pointsMap = {
  top: "4,10 8,6 12,10",
  right: "6,4 10,8 6,12",
  bottom: "4,6 8,10 12,6",
  left: "10,4 6,8 10,12"
};
var usePopoverDisclosureArrow = _RNZNGEL4cjs.createHook.call(void 0, 
  (_a) => {
    var _b = _a, { store, placement } = _b, props = _AV6KTKLEcjs.__objRest.call(void 0, _b, ["store", "placement"]);
    const context = _UVBBMANLcjs.usePopoverContext.call(void 0, );
    store = store || context;
    _misc.invariant.call(void 0, 
      store,
      process.env.NODE_ENV !== "production" && "PopoverDisclosureArrow must be wrapped in a PopoverDisclosure component."
    );
    const position = store.useState((state) => placement || state.placement);
    const dir = position.split("-")[0];
    const points = pointsMap[dir];
    const children = _react.useMemo.call(void 0, 
      () => /* @__PURE__ */ _jsxruntime.jsx.call(void 0, 
        "svg",
        {
          display: "block",
          fill: "none",
          stroke: "currentColor",
          strokeLinecap: "round",
          strokeLinejoin: "round",
          strokeWidth: "1.5pt",
          viewBox: "0 0 16 16",
          height: "1em",
          width: "1em",
          children: /* @__PURE__ */ _jsxruntime.jsx.call(void 0, "polyline", { points })
        }
      ),
      [points]
    );
    props = _AV6KTKLEcjs.__spreadProps.call(void 0, _AV6KTKLEcjs.__spreadValues.call(void 0, {
      children,
      "aria-hidden": true
    }, props), {
      style: _AV6KTKLEcjs.__spreadValues.call(void 0, {
        width: "1em",
        height: "1em",
        pointerEvents: "none"
      }, props.style)
    });
    return props;
  }
);
var PopoverDisclosureArrow = _RNZNGEL4cjs.createComponent.call(void 0, (props) => {
  const htmlProps = usePopoverDisclosureArrow(props);
  return _RNZNGEL4cjs.createElement.call(void 0, "span", htmlProps);
});
if (process.env.NODE_ENV !== "production") {
  PopoverDisclosureArrow.displayName = "PopoverDisclosureArrow";
}




exports.usePopoverDisclosureArrow = usePopoverDisclosureArrow; exports.PopoverDisclosureArrow = PopoverDisclosureArrow;
