// @wordpress/block-editor/build/components/block-tools/empty-block-inserter.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = EmptyBlockInserter;
var _react = require("react");
var _classnames = _interopRequireDefault(require("classnames"));
var _cover = _interopRequireDefault(require("../block-popover/cover"));
var _useBlockToolbarPopoverProps = _interopRequireDefault(require("./use-block-toolbar-popover-props"));
var _inserter = _interopRequireDefault(require("../inserter"));
var _useSelectedBlockToolProps = _interopRequireDefault(require("./use-selected-block-tool-props"));
/**
 * External dependencies
 */

/**
 * Internal dependencies
 */

function EmptyBlockInserter({
  clientId,
  __unstableContentRef
}) {
  const {
    capturingClientId,
    isInsertionPointVisible,
    lastClientId,
    rootClientId
  } = (0, _useSelectedBlockToolProps.default)(clientId);
  const popoverProps = (0, _useBlockToolbarPopoverProps.default)({
    contentElement: __unstableContentRef?.current,
    clientId
  });
  return (0, _react.createElement)(_cover.default, {
    clientId: capturingClientId || clientId,
    bottomClientId: lastClientId,
    className: (0, _classnames.default)('block-editor-block-list__block-side-inserter-popover', {
      'is-insertion-point-visible': isInsertionPointVisible
    }),
    __unstableContentRef: __unstableContentRef,
    ...popoverProps
  }, (0, _react.createElement)("div", {
    className: "block-editor-block-list__empty-block-inserter"
  }, (0, _react.createElement)(_inserter.default, {
    position: "bottom right",
    rootClientId: rootClientId,
    clientId: clientId,
    __experimentalIsQuick: true
  })));
}
