// @wordpress/components/build/progress-bar/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.ProgressBar = void 0;
var _react = require("react");
var _i18n = require("@wordpress/i18n");
var _element = require("@wordpress/element");
var ProgressBarStyled = _interopRequireWildcard(require("./styles"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && Object.prototype.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

function UnforwardedProgressBar(props, ref) {
  const {
    className,
    value,
    ...progressProps
  } = props;
  const isIndeterminate = !Number.isFinite(value);
  return (0, _react.createElement)(ProgressBarStyled.Track, {
    className: className
  }, (0, _react.createElement)(ProgressBarStyled.Indicator, {
    isIndeterminate: isIndeterminate,
    value: value
  }), (0, _react.createElement)(ProgressBarStyled.ProgressElement, {
    max: 100,
    value: value,
    "aria-label": (0, _i18n.__)('Loading …'),
    ref: ref,
    ...progressProps
  }));
}
const ProgressBar = exports.ProgressBar = (0, _element.forwardRef)(UnforwardedProgressBar);
var _default = exports.default = ProgressBar;
