// @wordpress/block-editor/build/components/inserter/library.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = require("react");
var _data = require("@wordpress/data");
var _element = require("@wordpress/element");
var _menu = _interopRequireDefault(require("./menu"));
var _store = require("../../store");
/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

const noop = () => {};
function InserterLibrary({
  rootClientId,
  clientId,
  isAppender,
  showInserterHelpPanel,
  showMostUsedBlocks = false,
  __experimentalInsertionIndex,
  __experimentalFilterValue,
  onSelect = noop,
  shouldFocusBlock = false
}, ref) {
  const {
    destinationRootClientId
  } = (0, _data.useSelect)(select => {
    const {
      getBlockRootClientId
    } = select(_store.store);
    const _rootClientId = rootClientId || getBlockRootClientId(clientId) || undefined;
    return {
      destinationRootClientId: _rootClientId
    };
  }, [clientId, rootClientId]);
  return (0, _react.createElement)(_menu.default, {
    onSelect: onSelect,
    rootClientId: destinationRootClientId,
    clientId: clientId,
    isAppender: isAppender,
    showInserterHelpPanel: showInserterHelpPanel,
    showMostUsedBlocks: showMostUsedBlocks,
    __experimentalInsertionIndex: __experimentalInsertionIndex,
    __experimentalFilterValue: __experimentalFilterValue,
    shouldFocusBlock: shouldFocusBlock,
    ref: ref
  });
}
var _default = exports.default = (0, _element.forwardRef)(InserterLibrary);
