// @wordpress/block-editor/build/components/inserter/reusable-blocks-tab.native.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = require("react");
var _data = require("@wordpress/data");
var _i18n = require("@wordpress/i18n");
var _element = require("@wordpress/element");
var _blockTypesList = _interopRequireDefault(require("../block-types-list"));
var _store = require("../../store");
var _utils = require("./utils");
/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

function ReusableBlocksTab({
  onSelect,
  rootClientId,
  listProps
}) {
  const {
    inserterItems
  } = (0, _data.useSelect)(select => {
    const {
      getInserterItems
    } = select(_store.store);
    const allItems = getInserterItems(rootClientId);
    return {
      inserterItems: allItems
    };
  }, [rootClientId]);
  const items = (0, _element.useMemo)(() => {
    return (0, _utils.filterInserterItems)(inserterItems, {
      onlyReusable: true
    });
  }, [inserterItems]);
  const sections = [(0, _utils.createInserterSection)({
    key: 'reuseable',
    items
  })];
  return (0, _react.createElement)(_blockTypesList.default, {
    name: "SyncedPatterns",
    sections: sections,
    onSelect: onSelect,
    listProps: listProps,
    label: (0, _i18n.__)('Synced patterns')
  });
}
var _default = exports.default = ReusableBlocksTab;
