// @wordpress/block-editor/build/components/inserter/tabs.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = require("react");
var _components = require("@wordpress/components");
var _i18n = require("@wordpress/i18n");
var _lockUnlock = require("../../lock-unlock");
/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

const {
  Tabs
} = (0, _lockUnlock.unlock)(_components.privateApis);
const blocksTab = {
  name: 'blocks',
  /* translators: Blocks tab title in the block inserter. */
  title: (0, _i18n.__)('Blocks')
};
const patternsTab = {
  name: 'patterns',
  /* translators: Theme and Directory Patterns tab title in the block inserter. */
  title: (0, _i18n.__)('Patterns')
};
const mediaTab = {
  name: 'media',
  /* translators: Media tab title in the block inserter. */
  title: (0, _i18n.__)('Media')
};
function InserterTabs({
  showBlocks = true,
  showPatterns = false,
  showMedia = false,
  onSelect,
  tabsContents
}) {
  const tabs = [showBlocks && blocksTab, showPatterns && patternsTab, showMedia && mediaTab].filter(Boolean);
  return (0, _react.createElement)("div", {
    className: "block-editor-inserter__tabs"
  }, (0, _react.createElement)(Tabs, {
    onSelect: onSelect
  }, (0, _react.createElement)(Tabs.TabList, null, tabs.map(tab => (0, _react.createElement)(Tabs.Tab, {
    key: tab.name,
    tabId: tab.name
  }, tab.title))), tabs.map(tab => (0, _react.createElement)(Tabs.TabPanel, {
    key: tab.name,
    tabId: tab.name,
    focusable: false
  }, tabsContents[tab.name]))));
}
var _default = exports.default = InserterTabs;
