// @wordpress/block-editor/build/components/inserter-listbox/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "InserterListboxGroup", {
  enumerable: true,
  get: function () {
    return _group.default;
  }
});
Object.defineProperty(exports, "InserterListboxItem", {
  enumerable: true,
  get: function () {
    return _item.default;
  }
});
Object.defineProperty(exports, "InserterListboxRow", {
  enumerable: true,
  get: function () {
    return _row.default;
  }
});
exports.default = void 0;
var _react = require("react");
var _components = require("@wordpress/components");
var _lockUnlock = require("../../lock-unlock");
var _group = _interopRequireDefault(require("./group"));
var _row = _interopRequireDefault(require("./row"));
var _item = _interopRequireDefault(require("./item"));
/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

const {
  CompositeV2: Composite,
  useCompositeStoreV2: useCompositeStore
} = (0, _lockUnlock.unlock)(_components.privateApis);
function InserterListbox({
  children
}) {
  const store = useCompositeStore({
    focusShift: true,
    focusWrap: 'horizontal'
  });
  return (0, _react.createElement)(Composite, {
    store: store,
    render: (0, _react.createElement)(_react.Fragment, null)
  }, children);
}
var _default = exports.default = InserterListbox;
