// @wordpress/block-editor/build/components/block-styles/menu-items.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = BlockStylesMenuItems;
var _react = require("react");
var _components = require("@wordpress/components");
var _icons = require("@wordpress/icons");
var _useStylesForBlock = _interopRequireDefault(require("./use-styles-for-block"));
/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

const noop = () => {};
function BlockStylesMenuItems({
  clientId,
  onSwitch = noop
}) {
  const {
    onSelect,
    stylesToRender,
    activeStyle
  } = (0, _useStylesForBlock.default)({
    clientId,
    onSwitch
  });
  if (!stylesToRender || stylesToRender.length === 0) {
    return null;
  }
  return (0, _react.createElement)(_react.Fragment, null, stylesToRender.map(style => {
    const menuItemText = style.label || style.name;
    return (0, _react.createElement)(_components.MenuItem, {
      key: style.name,
      icon: activeStyle.name === style.name ? _icons.check : null,
      onClick: () => onSelect(style)
    }, (0, _react.createElement)(_components.__experimentalText, {
      as: "span",
      limit: 18,
      ellipsizeMode: "tail",
      truncate: true
    }, menuItemText));
  }));
}
