// @wordpress/block-editor/build/components/link-control/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.URL_TYPE = exports.TEL_TYPE = exports.MAILTO_TYPE = exports.LINK_ENTRY_TYPES = exports.INTERNAL_TYPE = exports.DEFAULT_LINK_SETTINGS = exports.CREATE_TYPE = void 0;
var _i18n = require("@wordpress/i18n");
/**
 * WordPress dependencies
 */

// Used as a unique identifier for the "Create" option within search results.
// Used to help distinguish the "Create" suggestion within the search results in
// order to handle it as a unique case.
const CREATE_TYPE = exports.CREATE_TYPE = '__CREATE__';
const TEL_TYPE = exports.TEL_TYPE = 'tel';
const URL_TYPE = exports.URL_TYPE = 'link';
const MAILTO_TYPE = exports.MAILTO_TYPE = 'mailto';
const INTERNAL_TYPE = exports.INTERNAL_TYPE = 'internal';
const LINK_ENTRY_TYPES = exports.LINK_ENTRY_TYPES = [URL_TYPE, MAILTO_TYPE, TEL_TYPE, INTERNAL_TYPE];
const DEFAULT_LINK_SETTINGS = exports.DEFAULT_LINK_SETTINGS = [{
  id: 'opensInNewTab',
  title: (0, _i18n.__)('Open in new tab')
}];
