// @wordpress/block-editor/build-module/components/inserter-listbox/item.js
import { createElement } from "react";
/**
 * WordPress dependencies
 */
import { Button, privateApis as componentsPrivateApis } from '@wordpress/components';
import { forwardRef } from '@wordpress/element';

/**
 * Internal dependencies
 */
import { unlock } from '../../lock-unlock';
const {
  CompositeItemV2: CompositeItem
} = unlock(componentsPrivateApis);
function InserterListboxItem({
  isFirst,
  as: Component,
  children,
  ...props
}, ref) {
  return createElement(CompositeItem, {
    ref: ref,
    role: "option"
    // Use the CompositeItem `accessibleWhenDisabled` prop
    // over Button's `isFocusable`. The latter was shown to
    // cause an issue with tab order in the inserter list.
    ,
    accessibleWhenDisabled: true,
    ...props,
    render: htmlProps => {
      const propsWithTabIndex = {
        ...htmlProps,
        tabIndex: isFirst ? 0 : htmlProps.tabIndex
      };
      if (Component) {
        return createElement(Component, {
          ...propsWithTabIndex
        }, children);
      }
      if (typeof children === 'function') {
        return children(propsWithTabIndex);
      }
      return createElement(Button, {
        ...propsWithTabIndex
      }, children);
    }
  });
}
export default forwardRef(InserterListboxItem);
