// @wordpress/block-editor/build/components/inserter/block-patterns-explorer/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = require("react");
var _components = require("@wordpress/components");
var _element = require("@wordpress/element");
var _i18n = require("@wordpress/i18n");
var _patternExplorerSidebar = _interopRequireDefault(require("./pattern-explorer-sidebar"));
var _patternList = _interopRequireDefault(require("./pattern-list"));
var _usePatternCategories = require("../block-patterns-tab/use-pattern-categories");
/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

function PatternsExplorer({
  initialCategory,
  rootClientId
}) {
  const [searchValue, setSearchValue] = (0, _element.useState)('');
  const [patternSourceFilter, setPatternSourceFilter] = (0, _element.useState)('all');
  const [selectedCategory, setSelectedCategory] = (0, _element.useState)(initialCategory?.name);
  const patternCategories = (0, _usePatternCategories.usePatternCategories)(rootClientId, patternSourceFilter);
  return (0, _react.createElement)("div", {
    className: "block-editor-block-patterns-explorer"
  }, (0, _react.createElement)(_patternExplorerSidebar.default, {
    selectedCategory: selectedCategory,
    patternCategories: patternCategories,
    onClickCategory: setSelectedCategory,
    searchValue: searchValue,
    setSearchValue: setSearchValue,
    patternSourceFilter: patternSourceFilter,
    setPatternSourceFilter: setPatternSourceFilter
  }), (0, _react.createElement)(_patternList.default, {
    searchValue: searchValue,
    selectedCategory: selectedCategory,
    patternCategories: patternCategories,
    patternSourceFilter: patternSourceFilter
  }));
}
function PatternsExplorerModal({
  onModalClose,
  ...restProps
}) {
  return (0, _react.createElement)(_components.Modal, {
    title: (0, _i18n.__)('Patterns'),
    onRequestClose: onModalClose,
    isFullScreen: true
  }, (0, _react.createElement)(PatternsExplorer, {
    ...restProps
  }));
}
var _default = exports.default = PatternsExplorerModal;
