// @wordpress/block-editor/build/components/inserter/block-patterns-explorer/pattern-explorer-sidebar.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = require("react");
var _components = require("@wordpress/components");
var _i18n = require("@wordpress/i18n");
/**
 * WordPress dependencies
 */

function PatternCategoriesList({
  selectedCategory,
  patternCategories,
  onClickCategory
}) {
  const baseClassName = 'block-editor-block-patterns-explorer__sidebar';
  return (0, _react.createElement)("div", {
    className: `${baseClassName}__categories-list`
  }, patternCategories.map(({
    name,
    label
  }) => {
    return (0, _react.createElement)(_components.Button, {
      key: name,
      label: label,
      className: `${baseClassName}__categories-list__item`,
      isPressed: selectedCategory === name,
      onClick: () => {
        onClickCategory(name);
      }
    }, label);
  }));
}
function PatternsExplorerSearch({
  searchValue,
  setSearchValue
}) {
  const baseClassName = 'block-editor-block-patterns-explorer__search';
  return (0, _react.createElement)("div", {
    className: baseClassName
  }, (0, _react.createElement)(_components.SearchControl, {
    __nextHasNoMarginBottom: true,
    onChange: setSearchValue,
    value: searchValue,
    label: (0, _i18n.__)('Search for patterns'),
    placeholder: (0, _i18n.__)('Search')
  }));
}
function PatternExplorerSidebar({
  selectedCategory,
  patternCategories,
  onClickCategory,
  searchValue,
  setSearchValue
}) {
  const baseClassName = 'block-editor-block-patterns-explorer__sidebar';
  return (0, _react.createElement)("div", {
    className: baseClassName
  }, (0, _react.createElement)(PatternsExplorerSearch, {
    searchValue: searchValue,
    setSearchValue: setSearchValue
  }), !searchValue && (0, _react.createElement)(PatternCategoriesList, {
    selectedCategory: selectedCategory,
    patternCategories: patternCategories,
    onClickCategory: onClickCategory
  }));
}
var _default = exports.default = PatternExplorerSidebar;
