// @wordpress/block-editor/build/components/block-heading-level-dropdown/index.native.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = HeadingLevelDropdown;
var _react = require("react");
var _i18n = require("@wordpress/i18n");
var _components = require("@wordpress/components");
var _headingLevelIcon = _interopRequireDefault(require("./heading-level-icon"));
/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

const HEADING_LEVELS = [1, 2, 3, 4, 5, 6];

/** @typedef {import('react').ComponentType} ComponentType */

/**
 * HeadingLevelDropdown props.
 *
 * @typedef WPHeadingLevelDropdownProps
 *
 * @property {number}                 selectedLevel The chosen heading level.
 * @property {(newValue:number)=>any} onChange      Callback to run when
 *                                                  toolbar value is changed.
 */

/**
 * Dropdown for selecting a heading level (1 through 6).
 *
 * @param {WPHeadingLevelDropdownProps} props Component props.
 *
 * @return {ComponentType} The toolbar.
 */
function HeadingLevelDropdown({
  options = HEADING_LEVELS,
  value,
  onChange
}) {
  const createLevelControl = (targetLevel, currentLevel, onChangeCallback) => {
    const isActive = targetLevel === currentLevel;
    return {
      icon: (0, _react.createElement)(_headingLevelIcon.default, {
        level: targetLevel,
        isPressed: isActive
      }),
      // translators: %s: heading level e.g: "1", "2", "3"
      title: (0, _i18n.sprintf)((0, _i18n.__)('Heading %d'), targetLevel),
      isActive,
      onClick: () => onChangeCallback(targetLevel)
    };
  };
  return (0, _react.createElement)(_components.DropdownMenu, {
    icon: (0, _react.createElement)(_headingLevelIcon.default, {
      level: value
    }),
    controls: options.map(index => createLevelControl(index, value, onChange)),
    label: (0, _i18n.__)('Change level')
  });
}
