// @wordpress/block-editor/build/components/media-upload/constants.native.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.OPTION_WORDPRESS_MEDIA_LIBRARY = exports.OPTION_TAKE_VIDEO = exports.OPTION_TAKE_PHOTO = exports.OPTION_INSERT_FROM_URL = exports.MEDIA_TYPE_VIDEO = exports.MEDIA_TYPE_IMAGE = exports.MEDIA_TYPE_AUDIO = exports.MEDIA_TYPE_ANY = void 0;
var _i18n = require("@wordpress/i18n");
/**
 * WordPress dependencies
 */

const MEDIA_TYPE_IMAGE = exports.MEDIA_TYPE_IMAGE = 'image';
const MEDIA_TYPE_VIDEO = exports.MEDIA_TYPE_VIDEO = 'video';
const MEDIA_TYPE_AUDIO = exports.MEDIA_TYPE_AUDIO = 'audio';
const MEDIA_TYPE_ANY = exports.MEDIA_TYPE_ANY = 'any';
const OPTION_TAKE_VIDEO = exports.OPTION_TAKE_VIDEO = (0, _i18n.__)('Take a Video');
const OPTION_TAKE_PHOTO = exports.OPTION_TAKE_PHOTO = (0, _i18n.__)('Take a Photo');
const OPTION_INSERT_FROM_URL = exports.OPTION_INSERT_FROM_URL = (0, _i18n.__)('Insert from URL');
const OPTION_WORDPRESS_MEDIA_LIBRARY = exports.OPTION_WORDPRESS_MEDIA_LIBRARY = (0, _i18n.__)('WordPress Media Library');
