// @wordpress/block-editor/build/components/global-styles/image-settings-panel.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = ImageSettingsPanel;
exports.useHasImageSettingsPanel = useHasImageSettingsPanel;
var _react = require("react");
var _components = require("@wordpress/components");
var _i18n = require("@wordpress/i18n");
var _utils = require("./utils");
/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

function useHasImageSettingsPanel(name, value, inheritedValue) {
  // Note: If lightbox `value` exists, that means it was
  // defined via the the Global Styles UI and will NOT
  // be a boolean value or contain the `allowEditing` property,
  // so we should show the settings panel in those cases.
  return name === 'core/image' && inheritedValue?.lightbox?.allowEditing || !!value?.lightbox;
}
function ImageSettingsPanel({
  onChange,
  value,
  inheritedValue,
  panelId
}) {
  const resetLightbox = () => {
    onChange(undefined);
  };
  const onChangeLightbox = newSetting => {
    onChange({
      enabled: newSetting
    });
  };
  let lightboxChecked = false;
  if (inheritedValue?.lightbox?.enabled) {
    lightboxChecked = inheritedValue.lightbox.enabled;
  }
  return (0, _react.createElement)(_react.Fragment, null, (0, _react.createElement)(_components.__experimentalToolsPanel, {
    label: (0, _i18n._x)('Settings', 'Image settings'),
    resetAll: resetLightbox,
    panelId: panelId,
    dropdownMenuProps: _utils.TOOLSPANEL_DROPDOWNMENU_PROPS
  }, (0, _react.createElement)(_components.__experimentalToolsPanelItem
  // We use the `userSettings` prop instead of `settings`, because `settings`
  // contains the core/theme values for the lightbox and we want to show the
  // "RESET" button ONLY when the user has explicitly set a value in the
  // Global Styles.
  , {
    hasValue: () => !!value?.lightbox,
    label: (0, _i18n.__)('Expand on click'),
    onDeselect: resetLightbox,
    isShownByDefault: true,
    panelId: panelId
  }, (0, _react.createElement)(_components.ToggleControl, {
    label: (0, _i18n.__)('Expand on click'),
    checked: lightboxChecked,
    onChange: onChangeLightbox
  }))));
}
