// @wordpress/block-editor/build/components/global-styles/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "AdvancedPanel", {
  enumerable: true,
  get: function () {
    return _advancedPanel.default;
  }
});
Object.defineProperty(exports, "BorderPanel", {
  enumerable: true,
  get: function () {
    return _borderPanel.default;
  }
});
Object.defineProperty(exports, "ColorPanel", {
  enumerable: true,
  get: function () {
    return _colorPanel.default;
  }
});
Object.defineProperty(exports, "DimensionsPanel", {
  enumerable: true,
  get: function () {
    return _dimensionsPanel.default;
  }
});
Object.defineProperty(exports, "FiltersPanel", {
  enumerable: true,
  get: function () {
    return _filtersPanel.default;
  }
});
Object.defineProperty(exports, "GlobalStylesContext", {
  enumerable: true,
  get: function () {
    return _context.GlobalStylesContext;
  }
});
Object.defineProperty(exports, "ImageSettingsPanel", {
  enumerable: true,
  get: function () {
    return _imageSettingsPanel.default;
  }
});
Object.defineProperty(exports, "TypographyPanel", {
  enumerable: true,
  get: function () {
    return _typographyPanel.default;
  }
});
Object.defineProperty(exports, "areGlobalStyleConfigsEqual", {
  enumerable: true,
  get: function () {
    return _utils.areGlobalStyleConfigsEqual;
  }
});
Object.defineProperty(exports, "getBlockCSSSelector", {
  enumerable: true,
  get: function () {
    return _getBlockCssSelector.getBlockCSSSelector;
  }
});
Object.defineProperty(exports, "getGlobalStylesChanges", {
  enumerable: true,
  get: function () {
    return _getGlobalStylesChanges.default;
  }
});
Object.defineProperty(exports, "getLayoutStyles", {
  enumerable: true,
  get: function () {
    return _useGlobalStylesOutput.getLayoutStyles;
  }
});
Object.defineProperty(exports, "useGlobalSetting", {
  enumerable: true,
  get: function () {
    return _hooks.useGlobalSetting;
  }
});
Object.defineProperty(exports, "useGlobalStyle", {
  enumerable: true,
  get: function () {
    return _hooks.useGlobalStyle;
  }
});
Object.defineProperty(exports, "useGlobalStylesOutput", {
  enumerable: true,
  get: function () {
    return _useGlobalStylesOutput.useGlobalStylesOutput;
  }
});
Object.defineProperty(exports, "useGlobalStylesOutputWithConfig", {
  enumerable: true,
  get: function () {
    return _useGlobalStylesOutput.useGlobalStylesOutputWithConfig;
  }
});
Object.defineProperty(exports, "useGlobalStylesReset", {
  enumerable: true,
  get: function () {
    return _hooks.useGlobalStylesReset;
  }
});
Object.defineProperty(exports, "useHasBorderPanel", {
  enumerable: true,
  get: function () {
    return _borderPanel.useHasBorderPanel;
  }
});
Object.defineProperty(exports, "useHasBorderPanelControls", {
  enumerable: true,
  get: function () {
    return _borderPanel.useHasBorderPanelControls;
  }
});
Object.defineProperty(exports, "useHasColorPanel", {
  enumerable: true,
  get: function () {
    return _colorPanel.useHasColorPanel;
  }
});
Object.defineProperty(exports, "useHasDimensionsPanel", {
  enumerable: true,
  get: function () {
    return _dimensionsPanel.useHasDimensionsPanel;
  }
});
Object.defineProperty(exports, "useHasFiltersPanel", {
  enumerable: true,
  get: function () {
    return _filtersPanel.useHasFiltersPanel;
  }
});
Object.defineProperty(exports, "useHasImageSettingsPanel", {
  enumerable: true,
  get: function () {
    return _imageSettingsPanel.useHasImageSettingsPanel;
  }
});
Object.defineProperty(exports, "useHasTypographyPanel", {
  enumerable: true,
  get: function () {
    return _typographyPanel.useHasTypographyPanel;
  }
});
Object.defineProperty(exports, "useSettingsForBlockElement", {
  enumerable: true,
  get: function () {
    return _hooks.useSettingsForBlockElement;
  }
});
var _hooks = require("./hooks");
var _getBlockCssSelector = require("./get-block-css-selector");
var _useGlobalStylesOutput = require("./use-global-styles-output");
var _context = require("./context");
var _typographyPanel = _interopRequireWildcard(require("./typography-panel"));
var _dimensionsPanel = _interopRequireWildcard(require("./dimensions-panel"));
var _borderPanel = _interopRequireWildcard(require("./border-panel"));
var _colorPanel = _interopRequireWildcard(require("./color-panel"));
var _filtersPanel = _interopRequireWildcard(require("./filters-panel"));
var _imageSettingsPanel = _interopRequireWildcard(require("./image-settings-panel"));
var _advancedPanel = _interopRequireDefault(require("./advanced-panel"));
var _utils = require("./utils");
var _getGlobalStylesChanges = _interopRequireDefault(require("./get-global-styles-changes"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && Object.prototype.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
