// @wordpress/block-editor/build/components/global-styles/shadow-panel-components.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ShadowIndicator = ShadowIndicator;
exports.ShadowPopover = ShadowPopover;
exports.ShadowPopoverContainer = ShadowPopoverContainer;
exports.ShadowPresets = ShadowPresets;
exports.useShadowPresets = useShadowPresets;
var _react = require("react");
var _i18n = require("@wordpress/i18n");
var _components = require("@wordpress/components");
var _element = require("@wordpress/element");
var _icons = require("@wordpress/icons");
var _classnames = _interopRequireDefault(require("classnames"));
var _lockUnlock = require("../../lock-unlock");
/**
 * WordPress dependencies
 */

/**
 * External dependencies
 */

/**
 * Internal dependencies
 */

/**
 * Shared reference to an empty array for cases where it is important to avoid
 * returning a new array reference on every invocation.
 *
 * @type {Array}
 */
const EMPTY_ARRAY = [];
function ShadowPopoverContainer({
  shadow,
  onShadowChange,
  settings
}) {
  const shadows = useShadowPresets(settings);
  return (0, _react.createElement)("div", {
    className: "block-editor-global-styles__shadow-popover-container"
  }, (0, _react.createElement)(_components.__experimentalVStack, {
    spacing: 4
  }, (0, _react.createElement)(_components.__experimentalHeading, {
    level: 5
  }, (0, _i18n.__)('Drop shadow')), (0, _react.createElement)(ShadowPresets, {
    presets: shadows,
    activeShadow: shadow,
    onSelect: onShadowChange
  }), (0, _react.createElement)("div", {
    className: "block-editor-global-styles__clear-shadow"
  }, (0, _react.createElement)(_components.Button, {
    variant: "tertiary",
    onClick: () => onShadowChange(undefined)
  }, (0, _i18n.__)('Clear')))));
}
function ShadowPresets({
  presets,
  activeShadow,
  onSelect
}) {
  const {
    CompositeV2: Composite,
    useCompositeStoreV2: useCompositeStore
  } = (0, _lockUnlock.unlock)(_components.privateApis);
  const compositeStore = useCompositeStore();
  return !presets ? null : (0, _react.createElement)(Composite, {
    store: compositeStore,
    role: "listbox",
    className: "block-editor-global-styles__shadow__list",
    "aria-label": (0, _i18n.__)('Drop shadows')
  }, presets.map(({
    name,
    slug,
    shadow
  }) => (0, _react.createElement)(ShadowIndicator, {
    key: slug,
    label: name,
    isActive: shadow === activeShadow,
    type: slug === 'unset' ? 'unset' : 'preset',
    onSelect: () => onSelect(shadow === activeShadow ? undefined : shadow),
    shadow: shadow
  })));
}
function ShadowIndicator({
  type,
  label,
  isActive,
  onSelect,
  shadow
}) {
  const {
    CompositeItemV2: CompositeItem
  } = (0, _lockUnlock.unlock)(_components.privateApis);
  return (0, _react.createElement)(CompositeItem, {
    role: "option",
    "aria-label": label,
    "aria-selected": isActive,
    className: (0, _classnames.default)('block-editor-global-styles__shadow__item', {
      'is-active': isActive
    }),
    render: (0, _react.createElement)(_components.Button, {
      className: (0, _classnames.default)('block-editor-global-styles__shadow-indicator', {
        unset: type === 'unset'
      }),
      onClick: onSelect,
      label: label,
      style: {
        boxShadow: shadow
      },
      showTooltip: true
    }, isActive && (0, _react.createElement)(_icons.Icon, {
      icon: _icons.check
    }))
  });
}
function ShadowPopover({
  shadow,
  onShadowChange,
  settings
}) {
  const popoverProps = {
    placement: 'left-start',
    offset: 36,
    shift: true
  };
  return (0, _react.createElement)(_components.Dropdown, {
    popoverProps: popoverProps,
    className: "block-editor-global-styles__shadow-dropdown",
    renderToggle: renderShadowToggle(),
    renderContent: () => (0, _react.createElement)(_components.__experimentalDropdownContentWrapper, {
      paddingSize: "medium"
    }, (0, _react.createElement)(ShadowPopoverContainer, {
      shadow: shadow,
      onShadowChange: onShadowChange,
      settings: settings
    }))
  });
}
function renderShadowToggle() {
  return ({
    onToggle,
    isOpen
  }) => {
    const toggleProps = {
      onClick: onToggle,
      className: (0, _classnames.default)({
        'is-open': isOpen
      }),
      'aria-expanded': isOpen
    };
    return (0, _react.createElement)(_components.Button, {
      ...toggleProps
    }, (0, _react.createElement)(_components.__experimentalHStack, {
      justify: "flex-start"
    }, (0, _react.createElement)(_icons.Icon, {
      className: "block-editor-global-styles__toggle-icon",
      icon: _icons.shadow,
      size: 24
    }), (0, _react.createElement)(_components.FlexItem, null, (0, _i18n.__)('Drop shadow'))));
  };
}
function useShadowPresets(settings) {
  return (0, _element.useMemo)(() => {
    var _settings$shadow$pres;
    if (!settings?.shadow) {
      return EMPTY_ARRAY;
    }
    const defaultPresetsEnabled = settings?.shadow?.defaultPresets;
    const {
      default: defaultShadows,
      theme: themeShadows
    } = (_settings$shadow$pres = settings?.shadow?.presets) !== null && _settings$shadow$pres !== void 0 ? _settings$shadow$pres : {};
    const unsetShadow = {
      name: (0, _i18n.__)('Unset'),
      slug: 'unset',
      shadow: 'none'
    };
    const shadowPresets = [...(defaultPresetsEnabled && defaultShadows || EMPTY_ARRAY), ...(themeShadows || EMPTY_ARRAY)];
    if (shadowPresets.length) {
      shadowPresets.unshift(unsetShadow);
    }
    return shadowPresets;
  }, [settings]);
}
