// @wordpress/block-editor/build/components/block-popover/cover.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = require("react");
var _element = require("@wordpress/element");
var _useBlockRefs = require("../block-list/use-block-props/use-block-refs");
var _ = _interopRequireDefault(require("."));
/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

function BlockPopoverCover({
  clientId,
  bottomClientId,
  children,
  shift = false,
  ...props
}, ref) {
  var _bottomClientId;
  (_bottomClientId = bottomClientId) !== null && _bottomClientId !== void 0 ? _bottomClientId : bottomClientId = clientId;
  const selectedElement = (0, _useBlockRefs.__unstableUseBlockElement)(clientId);
  return (0, _react.createElement)(_.default, {
    ref: ref,
    clientId: clientId,
    bottomClientId: bottomClientId,
    shift: shift,
    ...props
  }, selectedElement && clientId === bottomClientId ? (0, _react.createElement)(CoverContainer, {
    selectedElement: selectedElement
  }, children) : children);
}
function CoverContainer({
  selectedElement,
  children
}) {
  const [width, setWidth] = (0, _element.useState)(selectedElement.offsetWidth);
  const [height, setHeight] = (0, _element.useState)(selectedElement.offsetHeight);
  (0, _element.useEffect)(() => {
    const observer = new window.ResizeObserver(() => {
      setWidth(selectedElement.offsetWidth);
      setHeight(selectedElement.offsetHeight);
    });
    observer.observe(selectedElement, {
      box: 'border-box'
    });
    return () => observer.disconnect();
  }, [selectedElement]);
  const style = (0, _element.useMemo)(() => {
    return {
      position: 'absolute',
      width,
      height
    };
  }, [width, height]);
  return (0, _react.createElement)("div", {
    style: style
  }, children);
}
var _default = exports.default = (0, _element.forwardRef)(BlockPopoverCover);
