// @wordpress/components/build/mobile/keyboard-aware-flat-list/index.android.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.KeyboardAwareFlatList = void 0;
var _react = require("react");
var _reactNative = require("react-native");
var _reactNativeReanimated = _interopRequireDefault(require("react-native-reanimated"));
var _element = require("@wordpress/element");
var _useScroll = _interopRequireDefault(require("./use-scroll"));
var _keyboardAvoidingView = _interopRequireDefault(require("../keyboard-avoiding-view"));
/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

const AnimatedFlatList = _reactNativeReanimated.default.createAnimatedComponent(_reactNative.FlatList);
const KeyboardAwareFlatList = ({
  onScroll,
  ...props
}, ref) => {
  const {
    extraScrollHeight,
    scrollEnabled,
    shouldPreventAutomaticScroll
  } = props;
  const {
    scrollViewRef,
    scrollHandler,
    scrollToSection,
    scrollToElement,
    onContentSizeChange,
    lastScrollTo
  } = (0, _useScroll.default)({
    scrollEnabled,
    shouldPreventAutomaticScroll,
    extraScrollHeight,
    onScroll
  });
  const getFlatListRef = (0, _element.useCallback)(flatListRef => {
    // On Android, we get the ref of the associated scroll
    // view to the FlatList.
    scrollViewRef.current = flatListRef?.getNativeScrollRef();
  }, [scrollViewRef]);
  (0, _element.useImperativeHandle)(ref, () => {
    return {
      scrollViewRef: scrollViewRef.current,
      scrollToSection,
      scrollToElement,
      lastScrollTo
    };
  });
  return (0, _react.createElement)(_keyboardAvoidingView.default, {
    style: {
      flex: 1
    }
  }, (0, _react.createElement)(AnimatedFlatList, {
    ref: getFlatListRef,
    onScroll: scrollHandler,
    onContentSizeChange: onContentSizeChange,
    ...props
  }));
};
exports.KeyboardAwareFlatList = KeyboardAwareFlatList;
var _default = exports.default = (0, _element.forwardRef)(KeyboardAwareFlatList);
