// @wordpress/block-editor/build-module/components/rich-text/format-toolbar-container.js
import { createElement } from "react";
/**
 * WordPress dependencies
 */
import { __ } from '@wordpress/i18n';
import { Popover, ToolbarGroup } from '@wordpress/components';

/**
 * Internal dependencies
 */
import BlockControls from '../block-controls';
import FormatToolbar from './format-toolbar';
import NavigableToolbar from '../navigable-toolbar';
function InlineToolbar({
  popoverAnchor
}) {
  return createElement(Popover, {
    placement: "top",
    focusOnMount: false,
    anchor: popoverAnchor,
    className: "block-editor-rich-text__inline-format-toolbar",
    __unstableSlotName: "block-toolbar"
  }, createElement(NavigableToolbar, {
    className: "block-editor-rich-text__inline-format-toolbar-group"
    /* translators: accessibility text for the inline format toolbar */,
    "aria-label": __('Format tools')
  }, createElement(ToolbarGroup, null, createElement(FormatToolbar, null))));
}
const FormatToolbarContainer = ({
  inline,
  editableContentElement
}) => {
  if (inline) {
    return createElement(InlineToolbar, {
      popoverAnchor: editableContentElement
    });
  }

  // Render regular toolbar.
  return createElement(BlockControls, {
    group: "inline"
  }, createElement(FormatToolbar, null));
};
export default FormatToolbarContainer;
