// @wordpress/block-editor/build/components/inserter/media-tab/media-list.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = require("react");
var _components = require("@wordpress/components");
var _i18n = require("@wordpress/i18n");
var _mediaPreview = require("./media-preview");
var _lockUnlock = require("../../../lock-unlock");
/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

const {
  CompositeV2: Composite,
  useCompositeStoreV2: useCompositeStore
} = (0, _lockUnlock.unlock)(_components.privateApis);
function MediaList({
  mediaList,
  category,
  onClick,
  label = (0, _i18n.__)('Media List')
}) {
  const compositeStore = useCompositeStore();
  return (0, _react.createElement)(Composite, {
    store: compositeStore,
    role: "listbox",
    className: "block-editor-inserter__media-list",
    "aria-label": label
  }, mediaList.map((media, index) => (0, _react.createElement)(_mediaPreview.MediaPreview, {
    key: media.id || media.sourceId || index,
    media: media,
    category: category,
    onClick: onClick
  })));
}
var _default = exports.default = MediaList;
