// @wordpress/block-editor/build-module/components/block-popover/cover.js
import { createElement } from "react";
/**
 * WordPress dependencies
 */
import { useEffect, useState, useMemo, forwardRef } from '@wordpress/element';

/**
 * Internal dependencies
 */
import { __unstableUseBlockElement as useBlockElement } from '../block-list/use-block-props/use-block-refs';
import BlockPopover from '.';
function BlockPopoverCover({
  clientId,
  bottomClientId,
  children,
  shift = false,
  ...props
}, ref) {
  var _bottomClientId;
  (_bottomClientId = bottomClientId) !== null && _bottomClientId !== void 0 ? _bottomClientId : bottomClientId = clientId;
  const selectedElement = useBlockElement(clientId);
  return createElement(BlockPopover, {
    ref: ref,
    clientId: clientId,
    bottomClientId: bottomClientId,
    shift: shift,
    ...props
  }, selectedElement && clientId === bottomClientId ? createElement(CoverContainer, {
    selectedElement: selectedElement
  }, children) : children);
}
function CoverContainer({
  selectedElement,
  children
}) {
  const [width, setWidth] = useState(selectedElement.offsetWidth);
  const [height, setHeight] = useState(selectedElement.offsetHeight);
  useEffect(() => {
    const observer = new window.ResizeObserver(() => {
      setWidth(selectedElement.offsetWidth);
      setHeight(selectedElement.offsetHeight);
    });
    observer.observe(selectedElement, {
      box: 'border-box'
    });
    return () => observer.disconnect();
  }, [selectedElement]);
  const style = useMemo(() => {
    return {
      position: 'absolute',
      width,
      height
    };
  }, [width, height]);
  return createElement("div", {
    style: style
  }, children);
}
export default forwardRef(BlockPopoverCover);
