// @wordpress/components/build/text-control/index.native.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = require("react");
var _element = require("@wordpress/element");
var _cell = _interopRequireDefault(require("../mobile/bottom-sheet/cell"));
/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

function TextControl({
  label,
  hideLabelFromVision,
  value,
  help,
  className,
  instanceId,
  onChange,
  type = 'text',
  placeholder,
  ...props
}) {
  const id = `inspector-text-control-${instanceId}`;
  return (0, _react.createElement)(_cell.default, {
    label: label,
    hideLabelFromVision: hideLabelFromVision,
    id: id,
    help: help,
    className: className,
    type: type,
    value: value,
    onChangeValue: onChange,
    "aria-describedby": !!help ? id + '__help' : undefined,
    valuePlaceholder: placeholder,
    ...props
  });
}
var _default = exports.default = (0, _element.memo)(TextControl);
