// @wordpress/block-editor/build/components/block-toolbar/index.native.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = BlockToolbar;
var _react = require("react");
var _data = require("@wordpress/data");
var _element = require("@wordpress/element");
var _blockToolbarMenu = _interopRequireDefault(require("./block-toolbar-menu"));
var _blockControls = _interopRequireDefault(require("../block-controls"));
var _blockMover = _interopRequireDefault(require("../block-mover"));
var _ungroupButton = _interopRequireDefault(require("../ungroup-button"));
var _blockSettings = require("../block-settings");
var _store = require("../../store");
/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

const REMOVE_EMPY_PARENT_BLOCKS = ['core/buttons', 'core/columns', 'core/social-links'];
function BlockToolbar({
  anchorNodeRef
}) {
  const {
    rootClientId,
    blockClientId,
    isSelected,
    isValidAndVisual,
    isStackedHorizontally,
    parentBlockName,
    parentNumberOfInnerBlocks
  } = (0, _data.useSelect)(select => {
    const {
      getBlockListSettings,
      getBlockMode,
      getBlockName,
      getBlockCount,
      getBlockRootClientId,
      getSelectedBlockClientIds,
      isBlockValid
    } = select(_store.store);
    const selectedBlockClientIds = getSelectedBlockClientIds();
    const selectedBlockClientId = selectedBlockClientIds[0];
    const blockRootClientId = getBlockRootClientId(selectedBlockClientId);
    const blockListSettings = getBlockListSettings(blockRootClientId);
    const orientation = blockListSettings?.orientation;
    const isBlockStackedHorizontally = orientation === 'horizontal';
    const parentName = getBlockName(blockRootClientId);
    const numberOfInnerBlocks = getBlockCount(blockRootClientId);
    return {
      rootClientId: blockRootClientId,
      blockClientId: selectedBlockClientId,
      isSelected: selectedBlockClientIds.length > 0,
      isStackedHorizontally: isBlockStackedHorizontally,
      parentBlockName: parentName,
      parentNumberOfInnerBlocks: numberOfInnerBlocks,
      isValidAndVisual: selectedBlockClientIds.length === 1 ? isBlockValid(selectedBlockClientIds[0]) && getBlockMode(selectedBlockClientIds[0]) === 'visual' : false
    };
  }, []);
  const {
    removeBlock
  } = (0, _data.useDispatch)(_store.store);
  const onRemove = (0, _element.useCallback)(() => {
    // Temp: remove parent block for specific cases where they don't
    // have inner blocks, ideally we should match the behavior as in
    // the Web editor and show a placeholder instead of removing the parent.
    if (REMOVE_EMPY_PARENT_BLOCKS.includes(parentBlockName) && parentNumberOfInnerBlocks === 1) {
      removeBlock(rootClientId);
      return;
    }
    removeBlock(blockClientId);
  }, [blockClientId, parentBlockName, parentNumberOfInnerBlocks, removeBlock, rootClientId]);
  if (!isSelected) {
    return null;
  }
  return (0, _react.createElement)(_react.Fragment, null, isValidAndVisual && (0, _react.createElement)(_react.Fragment, null, (0, _react.createElement)(_blockSettings.BlockSettingsButton.Slot, null, (fills = [null]) => {
    if (!fills?.length > 0) {
      return null;
    }
    return fills[0];
  }), (0, _react.createElement)(_blockControls.default.Slot, {
    group: "block"
  }), (0, _react.createElement)(_blockControls.default.Slot, null), (0, _react.createElement)(_blockControls.default.Slot, {
    group: "inline"
  }), (0, _react.createElement)(_blockControls.default.Slot, {
    group: "other"
  }), (0, _react.createElement)(_ungroupButton.default, null), (0, _react.createElement)(_blockMover.default, {
    clientIds: [blockClientId],
    isStackedHorizontally: isStackedHorizontally
  }), (0, _react.createElement)(_blockToolbarMenu.default, {
    clientId: blockClientId,
    isStackedHorizontally: isStackedHorizontally,
    onDelete: onRemove,
    anchorNodeRef: anchorNodeRef
  })));
}
