// @wordpress/block-editor/build/components/block-toolbar/shuffle.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = Shuffle;
var _react = require("react");
var _icons = require("@wordpress/icons");
var _components = require("@wordpress/components");
var _i18n = require("@wordpress/i18n");
var _element = require("@wordpress/element");
var _data = require("@wordpress/data");
var _store = require("../../store");
/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

const EMPTY_ARRAY = [];
function Container(props) {
  return (0, _react.createElement)(_components.ToolbarGroup, null, (0, _react.createElement)(_components.ToolbarButton, {
    ...props
  }));
}
function Shuffle({
  clientId,
  as = Container
}) {
  const {
    categories,
    patterns
  } = (0, _data.useSelect)(select => {
    const {
      getBlockAttributes,
      getBlockRootClientId,
      __experimentalGetAllowedPatterns
    } = select(_store.store);
    const attributes = getBlockAttributes(clientId);
    const _categories = attributes?.metadata?.categories || EMPTY_ARRAY;
    const rootBlock = getBlockRootClientId(clientId);
    const _patterns = __experimentalGetAllowedPatterns(rootBlock);
    return {
      categories: _categories,
      patterns: _patterns
    };
  }, [clientId]);
  const {
    replaceBlocks
  } = (0, _data.useDispatch)(_store.store);
  const sameCategoryPatternsWithSingleWrapper = (0, _element.useMemo)(() => {
    if (!categories || categories.length === 0 || !patterns || patterns.length === 0) {
      return EMPTY_ARRAY;
    }
    return patterns.filter(pattern => {
      return (
        // Check if the pattern has only one top level block,
        // otherwise we may shuffle to pattern that will not allow to continue shuffling.
        pattern.blocks.length === 1 && pattern.categories.some(category => {
          return categories.includes(category);
        })
      );
    });
  }, [categories, patterns]);
  if (sameCategoryPatternsWithSingleWrapper.length === 0) {
    return null;
  }
  const ComponentToUse = as;
  return (0, _react.createElement)(ComponentToUse, {
    label: (0, _i18n.__)('Shuffle'),
    icon: _icons.shuffle,
    onClick: () => {
      const randomPattern = sameCategoryPatternsWithSingleWrapper[Math.floor(
      // eslint-disable-next-line no-restricted-syntax
      Math.random() * sameCategoryPatternsWithSingleWrapper.length)];
      randomPattern.blocks[0].attributes = {
        ...randomPattern.blocks[0].attributes,
        metadata: {
          ...randomPattern.blocks[0].attributes.metadata,
          categories
        }
      };
      replaceBlocks(clientId, randomPattern.blocks);
    }
  });
}
