// @wordpress/block-editor/build-module/components/inspector-popover-header/index.js
import { createElement } from "react";
/**
 * WordPress dependencies
 */
import { __experimentalVStack as VStack, __experimentalHStack as HStack, __experimentalHeading as Heading, __experimentalSpacer as Spacer, Button, __experimentalText as Text } from '@wordpress/components';
import { closeSmall } from '@wordpress/icons';
import { __ } from '@wordpress/i18n';
export default function InspectorPopoverHeader({
  title,
  help,
  actions = [],
  onClose
}) {
  return createElement(VStack, {
    className: "block-editor-inspector-popover-header",
    spacing: 4
  }, createElement(HStack, {
    alignment: "center"
  }, createElement(Heading, {
    className: "block-editor-inspector-popover-header__heading",
    level: 2,
    size: 13
  }, title), createElement(Spacer, null), actions.map(({
    label,
    icon,
    onClick
  }) => createElement(Button, {
    key: label,
    className: "block-editor-inspector-popover-header__action",
    label: label,
    icon: icon,
    variant: !icon && 'tertiary',
    onClick: onClick
  }, !icon && label)), onClose && createElement(Button, {
    className: "block-editor-inspector-popover-header__action",
    label: __('Close'),
    icon: closeSmall,
    onClick: onClose
  })), help && createElement(Text, null, help));
}
