// @wordpress/components/build/radio-group/radio.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.Radio = void 0;
var _react = require("react");
var _element = require("@wordpress/element");
var Ariakit = _interopRequireWildcard(require("@ariakit/react"));
var _button = _interopRequireDefault(require("../button"));
var _context = require("./context");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && Object.prototype.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/**
 * WordPress dependencies
 */

/**
 * External dependencies
 */
// eslint-disable-next-line no-restricted-imports

/**
 * Internal dependencies
 */

function UnforwardedRadio({
  value,
  children,
  ...props
}, ref) {
  const {
    store,
    disabled
  } = (0, _element.useContext)(_context.RadioGroupContext);
  const selectedValue = store?.useState('value');
  const isChecked = selectedValue !== undefined && selectedValue === value;
  return (0, _react.createElement)(Ariakit.Radio, {
    disabled: disabled,
    store: store,
    ref: ref,
    value: value,
    render: (0, _react.createElement)(_button.default, {
      variant: isChecked ? 'primary' : 'secondary',
      ...props
    })
  }, children || value);
}

/**
 * @deprecated Use `RadioControl` or `ToggleGroupControl` instead.
 */
const Radio = exports.Radio = (0, _element.forwardRef)(UnforwardedRadio);
var _default = exports.default = Radio;
