// @wordpress/block-editor/build/components/warning/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = require("react");
var _classnames = _interopRequireDefault(require("classnames"));
var _element = require("@wordpress/element");
var _components = require("@wordpress/components");
var _i18n = require("@wordpress/i18n");
var _icons = require("@wordpress/icons");
/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */

function Warning({
  className,
  actions,
  children,
  secondaryActions
}) {
  return (0, _react.createElement)("div", {
    style: {
      display: 'contents',
      all: 'initial'
    }
  }, (0, _react.createElement)("div", {
    className: (0, _classnames.default)(className, 'block-editor-warning')
  }, (0, _react.createElement)("div", {
    className: "block-editor-warning__contents"
  }, (0, _react.createElement)("p", {
    className: "block-editor-warning__message"
  }, children), (_element.Children.count(actions) > 0 || secondaryActions) && (0, _react.createElement)("div", {
    className: "block-editor-warning__actions"
  }, _element.Children.count(actions) > 0 && _element.Children.map(actions, (action, i) => (0, _react.createElement)("span", {
    key: i,
    className: "block-editor-warning__action"
  }, action)), secondaryActions && (0, _react.createElement)(_components.DropdownMenu, {
    className: "block-editor-warning__secondary",
    icon: _icons.moreVertical,
    label: (0, _i18n.__)('More options'),
    popoverProps: {
      position: 'bottom left',
      className: 'block-editor-warning__dropdown'
    },
    noIcons: true
  }, () => (0, _react.createElement)(_components.MenuGroup, null, secondaryActions.map((item, pos) => (0, _react.createElement)(_components.MenuItem, {
    onClick: item.onClick,
    key: pos
  }, item.title))))))));
}

/**
 * @see https://github.com/WordPress/gutenberg/blob/HEAD/packages/block-editor/src/components/warning/README.md
 */
var _default = exports.default = Warning;
