// @wordpress/components/build/slot-fill/bubbles-virtually/slot-fill-context.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _utils = require("valtio/utils");
var _element = require("@wordpress/element");
var _warning = _interopRequireDefault(require("@wordpress/warning"));
/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */

const initialContextValue = {
  slots: (0, _utils.proxyMap)(),
  fills: (0, _utils.proxyMap)(),
  registerSlot: () => {
    typeof SCRIPT_DEBUG !== "undefined" && SCRIPT_DEBUG === true ? (0, _warning.default)('Components must be wrapped within `SlotFillProvider`. ' + 'See https://developer.wordpress.org/block-editor/components/slot-fill/') : void 0;
  },
  updateSlot: () => {},
  unregisterSlot: () => {},
  registerFill: () => {},
  unregisterFill: () => {},
  // This helps the provider know if it's using the default context value or not.
  isDefault: true
};
const SlotFillContext = (0, _element.createContext)(initialContextValue);
var _default = exports.default = SlotFillContext;
