// @wordpress/block-editor/build/components/block-patterns-paging/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = Pagination;
var _react = require("react");
var _components = require("@wordpress/components");
var _i18n = require("@wordpress/i18n");
/**
 * WordPress dependencies
 */

function Pagination({
  currentPage,
  numPages,
  changePage,
  totalItems
}) {
  return (0, _react.createElement)(_components.__experimentalVStack, {
    className: "block-editor-patterns__grid-pagination-wrapper"
  }, (0, _react.createElement)(_components.__experimentalText, {
    variant: "muted"
  },
  // translators: %s: Total number of patterns.
  (0, _i18n.sprintf)(
  // translators: %s: Total number of patterns.
  (0, _i18n._n)('%s item', '%s items', totalItems), totalItems)), numPages > 1 && (0, _react.createElement)(_components.__experimentalHStack, {
    expanded: false,
    spacing: 3,
    justify: "flex-start",
    className: "block-editor-patterns__grid-pagination"
  }, (0, _react.createElement)(_components.__experimentalHStack, {
    expanded: false,
    spacing: 1,
    className: "block-editor-patterns__grid-pagination-previous"
  }, (0, _react.createElement)(_components.Button, {
    variant: "tertiary",
    onClick: () => changePage(1),
    disabled: currentPage === 1,
    "aria-label": (0, _i18n.__)('First page')
  }, (0, _react.createElement)("span", null, "\xAB")), (0, _react.createElement)(_components.Button, {
    variant: "tertiary",
    onClick: () => changePage(currentPage - 1),
    disabled: currentPage === 1,
    "aria-label": (0, _i18n.__)('Previous page')
  }, (0, _react.createElement)("span", null, "\u2039"))), (0, _react.createElement)(_components.__experimentalText, {
    variant: "muted"
  }, (0, _i18n.sprintf)(
  // translators: %1$s: Current page number, %2$s: Total number of pages.
  (0, _i18n._x)('%1$s of %2$s', 'paging'), currentPage, numPages)), (0, _react.createElement)(_components.__experimentalHStack, {
    expanded: false,
    spacing: 1,
    className: "block-editor-patterns__grid-pagination-next"
  }, (0, _react.createElement)(_components.Button, {
    variant: "tertiary",
    onClick: () => changePage(currentPage + 1),
    disabled: currentPage === numPages,
    "aria-label": (0, _i18n.__)('Next page')
  }, (0, _react.createElement)("span", null, "\u203A")), (0, _react.createElement)(_components.Button, {
    variant: "tertiary",
    onClick: () => changePage(numPages),
    disabled: currentPage === numPages,
    "aria-label": (0, _i18n.__)('Last page'),
    size: "default"
  }, (0, _react.createElement)("span", null, "\xBB")))));
}
