// @wordpress/blocks/build/store/private-actions.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.addBootstrappedBlockType = addBootstrappedBlockType;
exports.addUnprocessedBlockType = addUnprocessedBlockType;
exports.registerBlockBindingsSource = registerBlockBindingsSource;
var _processBlockType = require("./process-block-type");
/**
 * Internal dependencies
 */

/** @typedef {import('../api/registration').WPBlockType} WPBlockType */

/**
 * Add bootstrapped block type metadata to the store. These metadata usually come from
 * the `block.json` file and are either statically boostrapped from the server, or
 * passed as the `metadata` parameter to the `registerBlockType` function.
 *
 * @param {string}      name      Block name.
 * @param {WPBlockType} blockType Block type metadata.
 */
function addBootstrappedBlockType(name, blockType) {
  return {
    type: 'ADD_BOOTSTRAPPED_BLOCK_TYPE',
    name,
    blockType
  };
}

/**
 * Add unprocessed block type settings to the store. These data are passed as the
 * `settings` parameter to the client-side `registerBlockType` function.
 *
 * @param {string}      name      Block name.
 * @param {WPBlockType} blockType Unprocessed block type settings.
 */
function addUnprocessedBlockType(name, blockType) {
  return ({
    dispatch
  }) => {
    dispatch({
      type: 'ADD_UNPROCESSED_BLOCK_TYPE',
      name,
      blockType
    });
    const processedBlockType = dispatch((0, _processBlockType.processBlockType)(name, blockType));
    if (!processedBlockType) {
      return;
    }
    dispatch.addBlockTypes(processedBlockType);
  };
}

/**
 * Register new block bindings source.
 *
 * @param {string} source Name of the source to register.
 */
function registerBlockBindingsSource(source) {
  return {
    type: 'REGISTER_BLOCK_BINDINGS_SOURCE',
    sourceName: source.name,
    sourceLabel: source.label,
    useSource: source.useSource,
    lockAttributesEditing: source.lockAttributesEditing
  };
}
