// @wordpress/block-editor/build/hooks/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "getBorderClassesAndStyles", {
  enumerable: true,
  get: function () {
    return _useBorderProps.getBorderClassesAndStyles;
  }
});
Object.defineProperty(exports, "getColorClassesAndStyles", {
  enumerable: true,
  get: function () {
    return _useColorProps.getColorClassesAndStyles;
  }
});
Object.defineProperty(exports, "getGapCSSValue", {
  enumerable: true,
  get: function () {
    return _gap.getGapCSSValue;
  }
});
Object.defineProperty(exports, "getShadowClassesAndStyles", {
  enumerable: true,
  get: function () {
    return _useShadowProps.getShadowClassesAndStyles;
  }
});
Object.defineProperty(exports, "getSpacingClassesAndStyles", {
  enumerable: true,
  get: function () {
    return _useSpacingProps.getSpacingClassesAndStyles;
  }
});
Object.defineProperty(exports, "getTypographyClassesAndStyles", {
  enumerable: true,
  get: function () {
    return _useTypographyProps.getTypographyClassesAndStyles;
  }
});
Object.defineProperty(exports, "useBorderProps", {
  enumerable: true,
  get: function () {
    return _useBorderProps.useBorderProps;
  }
});
Object.defineProperty(exports, "useCachedTruthy", {
  enumerable: true,
  get: function () {
    return _useCachedTruthy.useCachedTruthy;
  }
});
Object.defineProperty(exports, "useColorProps", {
  enumerable: true,
  get: function () {
    return _useColorProps.useColorProps;
  }
});
Object.defineProperty(exports, "useCustomSides", {
  enumerable: true,
  get: function () {
    return _dimensions.useCustomSides;
  }
});
Object.defineProperty(exports, "useLayoutClasses", {
  enumerable: true,
  get: function () {
    return _layout.useLayoutClasses;
  }
});
Object.defineProperty(exports, "useLayoutStyles", {
  enumerable: true,
  get: function () {
    return _layout.useLayoutStyles;
  }
});
var _utils = require("./utils");
require("./compat");
var _align = _interopRequireDefault(require("./align"));
require("./lock");
var _anchor = _interopRequireDefault(require("./anchor"));
var _ariaLabel = _interopRequireDefault(require("./aria-label"));
var _customClassName = _interopRequireDefault(require("./custom-class-name"));
require("./generated-class-name");
var _style = _interopRequireDefault(require("./style"));
require("./settings");
var _color = _interopRequireDefault(require("./color"));
var _dimensions = _interopRequireWildcard(require("./dimensions"));
var _duotone = _interopRequireDefault(require("./duotone"));
var _fontFamily = _interopRequireDefault(require("./font-family"));
var _fontSize = _interopRequireDefault(require("./font-size"));
var _border = _interopRequireDefault(require("./border"));
var _position = _interopRequireDefault(require("./position"));
var _layout = _interopRequireWildcard(require("./layout"));
var _layoutChild = _interopRequireDefault(require("./layout-child"));
var _contentLockUi = _interopRequireDefault(require("./content-lock-ui"));
require("./metadata");
var _blockHooks = _interopRequireDefault(require("./block-hooks"));
var _blockRenaming = _interopRequireDefault(require("./block-renaming"));
require("./use-bindings-attributes");
var _useBorderProps = require("./use-border-props");
var _useShadowProps = require("./use-shadow-props");
var _useColorProps = require("./use-color-props");
var _useSpacingProps = require("./use-spacing-props");
var _useTypographyProps = require("./use-typography-props");
var _gap = require("./gap");
var _useCachedTruthy = require("./use-cached-truthy");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && Object.prototype.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/**
 * Internal dependencies
 */

(0, _utils.createBlockEditFilter)([_align.default, _anchor.default, _customClassName.default, _style.default, _duotone.default, _position.default, _layout.default, _contentLockUi.default, _blockHooks.default, _blockRenaming.default, _layoutChild.default].filter(Boolean));
(0, _utils.createBlockListBlockFilter)([_align.default, _style.default, _color.default, _dimensions.default, _duotone.default, _fontFamily.default, _fontSize.default, _border.default, _position.default, _layoutChild.default]);
(0, _utils.createBlockSaveFilter)([_align.default, _anchor.default, _ariaLabel.default, _customClassName.default, _border.default, _color.default, _style.default, _fontFamily.default, _fontSize.default]);
