// @wordpress/block-editor/build/hooks/index.native.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "getBorderClassesAndStyles", {
  enumerable: true,
  get: function () {
    return _useBorderProps.getBorderClassesAndStyles;
  }
});
Object.defineProperty(exports, "getColorClassesAndStyles", {
  enumerable: true,
  get: function () {
    return _useColorProps.getColorClassesAndStyles;
  }
});
Object.defineProperty(exports, "getShadowClassesAndStyles", {
  enumerable: true,
  get: function () {
    return _useShadowProps.getShadowClassesAndStyles;
  }
});
Object.defineProperty(exports, "getSpacingClassesAndStyles", {
  enumerable: true,
  get: function () {
    return _useSpacingProps.getSpacingClassesAndStyles;
  }
});
Object.defineProperty(exports, "useBorderProps", {
  enumerable: true,
  get: function () {
    return _useBorderProps.useBorderProps;
  }
});
Object.defineProperty(exports, "useCachedTruthy", {
  enumerable: true,
  get: function () {
    return _useCachedTruthy.useCachedTruthy;
  }
});
Object.defineProperty(exports, "useColorProps", {
  enumerable: true,
  get: function () {
    return _useColorProps.useColorProps;
  }
});
Object.defineProperty(exports, "useEditorWrapperStyles", {
  enumerable: true,
  get: function () {
    return _useEditorWrapperStyles.useEditorWrapperStyles;
  }
});
var _utils = require("./utils");
require("./compat");
var _align = _interopRequireDefault(require("./align"));
var _anchor = _interopRequireDefault(require("./anchor"));
var _customClassName = _interopRequireDefault(require("./custom-class-name"));
require("./generated-class-name");
var _style = _interopRequireDefault(require("./style"));
var _color = _interopRequireDefault(require("./color"));
var _fontSize = _interopRequireDefault(require("./font-size"));
require("./layout");
var _useBorderProps = require("./use-border-props");
var _useShadowProps = require("./use-shadow-props");
var _useColorProps = require("./use-color-props");
var _useSpacingProps = require("./use-spacing-props");
var _useCachedTruthy = require("./use-cached-truthy");
var _useEditorWrapperStyles = require("./use-editor-wrapper-styles");
/**
 * Internal dependencies
 */

(0, _utils.createBlockEditFilter)([_align.default, _anchor.default, _style.default]);
(0, _utils.createBlockListBlockFilter)([_align.default, _style.default, _color.default, _fontSize.default]);
(0, _utils.createBlockSaveFilter)([_align.default, _anchor.default, _customClassName.default, _color.default, _style.default, _fontSize.default]);
