// @wordpress/block-editor/build/components/inspector-controls/fill.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = InspectorControlsFill;
var _react = require("react");
var _components = require("@wordpress/components");
var _warning = _interopRequireDefault(require("@wordpress/warning"));
var _deprecated = _interopRequireDefault(require("@wordpress/deprecated"));
var _element = require("@wordpress/element");
var _context = require("../block-edit/context");
var _groups = _interopRequireDefault(require("./groups"));
/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

function InspectorControlsFill({
  children,
  group = 'default',
  __experimentalGroup,
  resetAllFilter
}) {
  if (__experimentalGroup) {
    (0, _deprecated.default)('`__experimentalGroup` property in `InspectorControlsFill`', {
      since: '6.2',
      version: '6.4',
      alternative: '`group`'
    });
    group = __experimentalGroup;
  }
  const context = (0, _context.useBlockEditContext)();
  const Fill = _groups.default[group]?.Fill;
  if (!Fill) {
    typeof SCRIPT_DEBUG !== "undefined" && SCRIPT_DEBUG === true ? (0, _warning.default)(`Unknown InspectorControls group "${group}" provided.`) : void 0;
    return null;
  }
  if (!context[_context.mayDisplayControlsKey]) {
    return null;
  }
  return (0, _react.createElement)(_components.__experimentalStyleProvider, {
    document: document
  }, (0, _react.createElement)(Fill, null, fillProps => {
    return (0, _react.createElement)(ToolsPanelInspectorControl, {
      fillProps: fillProps,
      children: children,
      resetAllFilter: resetAllFilter
    });
  }));
}
function RegisterResetAll({
  resetAllFilter,
  children
}) {
  const {
    registerResetAllFilter,
    deregisterResetAllFilter
  } = (0, _element.useContext)(_components.__experimentalToolsPanelContext);
  (0, _element.useEffect)(() => {
    if (resetAllFilter && registerResetAllFilter && deregisterResetAllFilter) {
      registerResetAllFilter(resetAllFilter);
      return () => {
        deregisterResetAllFilter(resetAllFilter);
      };
    }
  }, [resetAllFilter, registerResetAllFilter, deregisterResetAllFilter]);
  return children;
}
function ToolsPanelInspectorControl({
  children,
  resetAllFilter,
  fillProps
}) {
  // `fillProps.forwardedContext` is an array of context provider entries, provided by slot,
  // that should wrap the fill markup.
  const {
    forwardedContext = []
  } = fillProps;

  // Children passed to InspectorControlsFill will not have
  // access to any React Context whose Provider is part of
  // the InspectorControlsSlot tree. So we re-create the
  // Provider in this subtree.
  const innerMarkup = (0, _react.createElement)(RegisterResetAll, {
    resetAllFilter: resetAllFilter
  }, children);
  return forwardedContext.reduce((inner, [Provider, props]) => (0, _react.createElement)(Provider, {
    ...props
  }, inner), innerMarkup);
}
