// @wordpress/block-editor/build/components/inspector-controls/groups.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _components = require("@wordpress/components");
/**
 * WordPress dependencies
 */

const InspectorControlsDefault = (0, _components.createSlotFill)('InspectorControls');
const InspectorControlsAdvanced = (0, _components.createSlotFill)('InspectorAdvancedControls');
const InspectorControlsBackground = (0, _components.createSlotFill)('InspectorControlsBackground');
const InspectorControlsBorder = (0, _components.createSlotFill)('InspectorControlsBorder');
const InspectorControlsColor = (0, _components.createSlotFill)('InspectorControlsColor');
const InspectorControlsFilter = (0, _components.createSlotFill)('InspectorControlsFilter');
const InspectorControlsDimensions = (0, _components.createSlotFill)('InspectorControlsDimensions');
const InspectorControlsPosition = (0, _components.createSlotFill)('InspectorControlsPosition');
const InspectorControlsTypography = (0, _components.createSlotFill)('InspectorControlsTypography');
const InspectorControlsListView = (0, _components.createSlotFill)('InspectorControlsListView');
const InspectorControlsStyles = (0, _components.createSlotFill)('InspectorControlsStyles');
const InspectorControlsEffects = (0, _components.createSlotFill)('InspectorControlsEffects');
const groups = {
  default: InspectorControlsDefault,
  advanced: InspectorControlsAdvanced,
  background: InspectorControlsBackground,
  border: InspectorControlsBorder,
  color: InspectorControlsColor,
  dimensions: InspectorControlsDimensions,
  effects: InspectorControlsEffects,
  filter: InspectorControlsFilter,
  list: InspectorControlsListView,
  position: InspectorControlsPosition,
  settings: InspectorControlsDefault,
  // Alias for default.
  styles: InspectorControlsStyles,
  typography: InspectorControlsTypography
};
var _default = exports.default = groups;
