// @wordpress/block-editor/build/components/inspector-controls/slot.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = InspectorControlsSlot;
var _react = require("react");
var _components = require("@wordpress/components");
var _element = require("@wordpress/element");
var _warning = _interopRequireDefault(require("@wordpress/warning"));
var _deprecated = _interopRequireDefault(require("@wordpress/deprecated"));
var _blockSupportToolsPanel = _interopRequireDefault(require("./block-support-tools-panel"));
var _blockSupportSlotContainer = _interopRequireDefault(require("./block-support-slot-container"));
var _groups = _interopRequireDefault(require("./groups"));
/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

function InspectorControlsSlot({
  __experimentalGroup,
  group = 'default',
  label,
  fillProps,
  ...props
}) {
  if (__experimentalGroup) {
    (0, _deprecated.default)('`__experimentalGroup` property in `InspectorControlsSlot`', {
      since: '6.2',
      version: '6.4',
      alternative: '`group`'
    });
    group = __experimentalGroup;
  }
  const Slot = _groups.default[group]?.Slot;
  const fills = (0, _components.__experimentalUseSlotFills)(Slot?.__unstableName);
  const motionContextValue = (0, _element.useContext)(_components.__unstableMotionContext);
  const computedFillProps = (0, _element.useMemo)(() => {
    var _fillProps$forwardedC;
    return {
      ...(fillProps !== null && fillProps !== void 0 ? fillProps : {}),
      forwardedContext: [...((_fillProps$forwardedC = fillProps?.forwardedContext) !== null && _fillProps$forwardedC !== void 0 ? _fillProps$forwardedC : []), [_components.__unstableMotionContext.Provider, {
        value: motionContextValue
      }]]
    };
  }, [motionContextValue, fillProps]);
  if (!Slot) {
    typeof SCRIPT_DEBUG !== "undefined" && SCRIPT_DEBUG === true ? (0, _warning.default)(`Unknown InspectorControls group "${group}" provided.`) : void 0;
    return null;
  }
  if (!fills?.length) {
    return null;
  }
  if (label) {
    return (0, _react.createElement)(_blockSupportToolsPanel.default, {
      group: group,
      label: label
    }, (0, _react.createElement)(_blockSupportSlotContainer.default, {
      ...props,
      fillProps: computedFillProps,
      Slot: Slot
    }));
  }
  return (0, _react.createElement)(Slot, {
    ...props,
    fillProps: computedFillProps,
    bubblesVirtually: true
  });
}
