// @wordpress/block-editor/build/components/url-popover/link-viewer.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = LinkViewer;
var _react = require("react");
var _classnames = _interopRequireDefault(require("classnames"));
var _i18n = require("@wordpress/i18n");
var _components = require("@wordpress/components");
var _icons = require("@wordpress/icons");
var _linkViewerUrl = _interopRequireDefault(require("./link-viewer-url"));
/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

function LinkViewer({
  className,
  linkClassName,
  onEditLinkClick,
  url,
  urlLabel,
  ...props
}) {
  return (0, _react.createElement)("div", {
    className: (0, _classnames.default)('block-editor-url-popover__link-viewer', className),
    ...props
  }, (0, _react.createElement)(_linkViewerUrl.default, {
    url: url,
    urlLabel: urlLabel,
    className: linkClassName
  }), onEditLinkClick && (0, _react.createElement)(_components.Button, {
    icon: _icons.edit,
    label: (0, _i18n.__)('Edit'),
    onClick: onEditLinkClick,
    size: "compact"
  }));
}
