// @wordpress/block-editor/build-module/components/inserter/tabs.js
import { createElement } from "react";
/**
 * WordPress dependencies
 */
import { privateApis as componentsPrivateApis } from '@wordpress/components';
import { __ } from '@wordpress/i18n';

/**
 * Internal dependencies
 */
import { unlock } from '../../lock-unlock';
const {
  Tabs
} = unlock(componentsPrivateApis);
const blocksTab = {
  name: 'blocks',
  /* translators: Blocks tab title in the block inserter. */
  title: __('Blocks')
};
const patternsTab = {
  name: 'patterns',
  /* translators: Theme and Directory Patterns tab title in the block inserter. */
  title: __('Patterns')
};
const mediaTab = {
  name: 'media',
  /* translators: Media tab title in the block inserter. */
  title: __('Media')
};
function InserterTabs({
  showBlocks = true,
  showPatterns = false,
  showMedia = false,
  onSelect,
  tabsContents
}) {
  const tabs = [showBlocks && blocksTab, showPatterns && patternsTab, showMedia && mediaTab].filter(Boolean);
  return createElement("div", {
    className: "block-editor-inserter__tabs"
  }, createElement(Tabs, {
    onSelect: onSelect
  }, createElement(Tabs.TabList, null, tabs.map(tab => createElement(Tabs.Tab, {
    key: tab.name,
    tabId: tab.name
  }, tab.title))), tabs.map(tab => createElement(Tabs.TabPanel, {
    key: tab.name,
    tabId: tab.name,
    focusable: false
  }, tabsContents[tab.name]))));
}
export default InserterTabs;
