// @wordpress/block-editor/build/components/list-view/leaf.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = require("react");
var _web = require("@react-spring/web");
var _classnames = _interopRequireDefault(require("classnames"));
var _components = require("@wordpress/components");
var _compose = require("@wordpress/compose");
var _element = require("@wordpress/element");
var _useMovingAnimation = _interopRequireDefault(require("../use-moving-animation"));
/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

const AnimatedTreeGridRow = (0, _web.animated)(_components.__experimentalTreeGridRow);
const ListViewLeaf = (0, _element.forwardRef)(({
  isDragged,
  isSelected,
  position,
  level,
  rowCount,
  children,
  className,
  path,
  ...props
}, ref) => {
  const animationRef = (0, _useMovingAnimation.default)({
    clientId: props['data-block'],
    enableAnimation: true,
    triggerAnimationOnChange: path
  });
  const mergedRef = (0, _compose.useMergeRefs)([ref, animationRef]);
  return (0, _react.createElement)(AnimatedTreeGridRow, {
    ref: mergedRef,
    className: (0, _classnames.default)('block-editor-list-view-leaf', className),
    level: level,
    positionInSet: position,
    setSize: rowCount,
    isExpanded: undefined,
    ...props
  }, children);
});
var _default = exports.default = ListViewLeaf;
