// @wordpress/block-editor/build-module/components/block-patterns-paging/index.js
import { createElement } from "react";
/**
 * WordPress dependencies
 */
import { __experimentalVStack as VStack, __experimentalHStack as HStack, __experimentalText as Text, Button } from '@wordpress/components';
import { __, _x, _n, sprintf } from '@wordpress/i18n';
export default function Pagination({
  currentPage,
  numPages,
  changePage,
  totalItems
}) {
  return createElement(VStack, {
    className: "block-editor-patterns__grid-pagination-wrapper"
  }, createElement(Text, {
    variant: "muted"
  },
  // translators: %s: Total number of patterns.
  sprintf(
  // translators: %s: Total number of patterns.
  _n('%s item', '%s items', totalItems), totalItems)), numPages > 1 && createElement(HStack, {
    expanded: false,
    spacing: 3,
    justify: "flex-start",
    className: "block-editor-patterns__grid-pagination"
  }, createElement(HStack, {
    expanded: false,
    spacing: 1,
    className: "block-editor-patterns__grid-pagination-previous"
  }, createElement(Button, {
    variant: "tertiary",
    onClick: () => changePage(1),
    disabled: currentPage === 1,
    "aria-label": __('First page')
  }, createElement("span", null, "\xAB")), createElement(Button, {
    variant: "tertiary",
    onClick: () => changePage(currentPage - 1),
    disabled: currentPage === 1,
    "aria-label": __('Previous page')
  }, createElement("span", null, "\u2039"))), createElement(Text, {
    variant: "muted"
  }, sprintf(
  // translators: %1$s: Current page number, %2$s: Total number of pages.
  _x('%1$s of %2$s', 'paging'), currentPage, numPages)), createElement(HStack, {
    expanded: false,
    spacing: 1,
    className: "block-editor-patterns__grid-pagination-next"
  }, createElement(Button, {
    variant: "tertiary",
    onClick: () => changePage(currentPage + 1),
    disabled: currentPage === numPages,
    "aria-label": __('Next page')
  }, createElement("span", null, "\u203A")), createElement(Button, {
    variant: "tertiary",
    onClick: () => changePage(numPages),
    disabled: currentPage === numPages,
    "aria-label": __('Last page'),
    size: "default"
  }, createElement("span", null, "\xBB")))));
}
