// @wordpress/block-editor/build-module/components/block-tools/block-toolbar-breadcrumb.js
import { createElement } from "react";
/**
 * External dependencies
 */
import classnames from 'classnames';

/**
 * Internal dependencies
 */
import BlockSelectionButton from './block-selection-button';
import BlockPopover from '../block-popover';
import useBlockToolbarPopoverProps from './use-block-toolbar-popover-props';
import useSelectedBlockToolProps from './use-selected-block-tool-props';
export default function BlockToolbarBreadcrumb({
  clientId,
  __unstableContentRef
}) {
  const {
    capturingClientId,
    isInsertionPointVisible,
    lastClientId,
    rootClientId
  } = useSelectedBlockToolProps(clientId);
  const popoverProps = useBlockToolbarPopoverProps({
    contentElement: __unstableContentRef?.current,
    clientId
  });
  return createElement(BlockPopover, {
    clientId: capturingClientId || clientId,
    bottomClientId: lastClientId,
    className: classnames('block-editor-block-list__block-popover', {
      'is-insertion-point-visible': isInsertionPointVisible
    }),
    resize: false,
    ...popoverProps
  }, createElement(BlockSelectionButton, {
    clientId: clientId,
    rootClientId: rootClientId
  }));
}
