// @wordpress/block-editor/build/components/block-quick-navigation/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = BlockQuickNavigation;
var _react = require("react");
var _data = require("@wordpress/data");
var _components = require("@wordpress/components");
var _blocks = require("@wordpress/blocks");
var _store = require("../../store");
var _blockIcon = _interopRequireDefault(require("../block-icon"));
/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

function BlockQuickNavigation({
  clientIds
}) {
  if (!clientIds.length) {
    return null;
  }
  return (0, _react.createElement)(_components.__experimentalVStack, {
    spacing: 1
  }, clientIds.map(clientId => (0, _react.createElement)(BlockQuickNavigationItem, {
    key: clientId,
    clientId: clientId
  })));
}
function BlockQuickNavigationItem({
  clientId
}) {
  const {
    name,
    icon,
    isSelected
  } = (0, _data.useSelect)(select => {
    const {
      getBlockName,
      getBlockAttributes,
      isBlockSelected,
      hasSelectedInnerBlock
    } = select(_store.store);
    const {
      getBlockType
    } = select(_blocks.store);
    const blockType = getBlockType(getBlockName(clientId));
    const attributes = getBlockAttributes(clientId);
    return {
      name: blockType && (0, _blocks.__experimentalGetBlockLabel)(blockType, attributes, 'list-view'),
      icon: blockType?.icon,
      isSelected: isBlockSelected(clientId) || hasSelectedInnerBlock(clientId, /* deep: */true)
    };
  }, [clientId]);
  const {
    selectBlock
  } = (0, _data.useDispatch)(_store.store);
  return (0, _react.createElement)(_components.Button, {
    isPressed: isSelected,
    onClick: () => selectBlock(clientId)
  }, (0, _react.createElement)(_components.Flex, null, (0, _react.createElement)(_components.FlexItem, null, (0, _react.createElement)(_blockIcon.default, {
    icon: icon
  })), (0, _react.createElement)(_components.FlexBlock, {
    style: {
      textAlign: 'left'
    }
  }, (0, _react.createElement)(_components.__experimentalTruncate, null, name))));
}
