// @wordpress/block-editor/build/components/inserter/block-patterns-tab/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = require("react");
var _element = require("@wordpress/element");
var _i18n = require("@wordpress/i18n");
var _compose = require("@wordpress/compose");
var _components = require("@wordpress/components");
var _icons = require("@wordpress/icons");
var _blockPatternsExplorer = _interopRequireDefault(require("../block-patterns-explorer"));
var _mobileTabNavigation = _interopRequireDefault(require("../mobile-tab-navigation"));
var _patternCategoryPreviews = require("./pattern-category-previews");
var _usePatternCategories = require("./use-pattern-categories");
/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

function BlockPatternsTab({
  onSelectCategory,
  selectedCategory,
  onInsert,
  rootClientId
}) {
  const [showPatternsExplorer, setShowPatternsExplorer] = (0, _element.useState)(false);
  const categories = (0, _usePatternCategories.usePatternCategories)(rootClientId);
  const initialCategory = selectedCategory || categories[0];
  const isMobile = (0, _compose.useViewportMatch)('medium', '<');
  return (0, _react.createElement)(_react.Fragment, null, !isMobile && (0, _react.createElement)("div", {
    className: "block-editor-inserter__block-patterns-tabs-container"
  }, (0, _react.createElement)("nav", {
    "aria-label": (0, _i18n.__)('Block pattern categories'),
    className: "block-editor-inserter__block-patterns-tabs"
  }, (0, _react.createElement)(_components.__experimentalItemGroup, {
    role: "list"
  }, categories.map(category => (0, _react.createElement)(_components.__experimentalItem, {
    role: "listitem",
    key: category.name,
    onClick: () => onSelectCategory(category),
    className: category === selectedCategory ? 'block-editor-inserter__patterns-category block-editor-inserter__patterns-selected-category' : 'block-editor-inserter__patterns-category',
    "aria-label": category.label,
    "aria-current": category === selectedCategory ? 'true' : undefined
  }, (0, _react.createElement)(_components.__experimentalHStack, null, (0, _react.createElement)(_components.FlexBlock, null, category.label), (0, _react.createElement)(_icons.Icon, {
    icon: (0, _i18n.isRTL)() ? _icons.chevronLeft : _icons.chevronRight
  })))), (0, _react.createElement)("div", {
    role: "listitem"
  }, (0, _react.createElement)(_components.Button, {
    className: "block-editor-inserter__patterns-explore-button",
    onClick: () => setShowPatternsExplorer(true),
    variant: "secondary"
  }, (0, _i18n.__)('Explore all patterns')))))), isMobile && (0, _react.createElement)(_mobileTabNavigation.default, {
    categories: categories
  }, category => (0, _react.createElement)(_patternCategoryPreviews.PatternCategoryPreviews, {
    key: category.name,
    onInsert: onInsert,
    rootClientId: rootClientId,
    category: category,
    showTitlesAsTooltip: false
  })), showPatternsExplorer && (0, _react.createElement)(_blockPatternsExplorer.default, {
    initialCategory: initialCategory,
    patternCategories: categories,
    onModalClose: () => setShowPatternsExplorer(false),
    rootClientId: rootClientId
  }));
}
var _default = exports.default = BlockPatternsTab;
