// @wordpress/block-editor/build/components/inserter/block-patterns-tab/pattern-category-preview-panel.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PatternCategoryPreviewPanel = PatternCategoryPreviewPanel;
var _react = require("react");
var _element = require("@wordpress/element");
var _dom = require("@wordpress/dom");
var _patternCategoryPreviews = require("./pattern-category-previews");
/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

function PatternCategoryPreviewPanel({
  rootClientId,
  onInsert,
  onHover,
  category,
  showTitlesAsTooltip,
  patternFilter
}) {
  const container = (0, _element.useRef)();
  (0, _element.useEffect)(() => {
    const timeout = setTimeout(() => {
      const [firstTabbable] = _dom.focus.tabbable.find(container.current);
      firstTabbable?.focus();
    });
    return () => clearTimeout(timeout);
  }, [category]);
  return (0, _react.createElement)("div", {
    ref: container,
    className: "block-editor-inserter__patterns-category-dialog"
  }, (0, _react.createElement)(_patternCategoryPreviews.PatternCategoryPreviews, {
    key: category.name,
    rootClientId: rootClientId,
    onInsert: onInsert,
    onHover: onHover,
    category: category,
    showTitlesAsTooltip: showTitlesAsTooltip,
    patternFilter: patternFilter
  }));
}
