// @wordpress/block-editor/build/components/inserter/block-patterns-tab/pattern-category-previews.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PatternCategoryPreviews = PatternCategoryPreviews;
var _react = require("react");
var _element = require("@wordpress/element");
var _i18n = require("@wordpress/i18n");
var _components = require("@wordpress/components");
var _usePatternsState = _interopRequireDefault(require("../hooks/use-patterns-state"));
var _blockPatternsList = _interopRequireDefault(require("../../block-patterns-list"));
var _usePatternsPaging = _interopRequireDefault(require("../hooks/use-patterns-paging"));
var _patternsFilter = require("./patterns-filter");
var _usePatternCategories = require("./use-pattern-categories");
var _utils = require("./utils");
/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

const noop = () => {};
function PatternCategoryPreviews({
  rootClientId,
  onInsert,
  onHover = noop,
  category,
  showTitlesAsTooltip
}) {
  const [allPatterns,, onClickPattern] = (0, _usePatternsState.default)(onInsert, rootClientId);
  const [patternSyncFilter, setPatternSyncFilter] = (0, _element.useState)('all');
  const [patternSourceFilter, setPatternSourceFilter] = (0, _element.useState)('all');
  const availableCategories = (0, _usePatternCategories.usePatternCategories)(rootClientId, patternSourceFilter);
  const scrollContainerRef = (0, _element.useRef)();
  const currentCategoryPatterns = (0, _element.useMemo)(() => allPatterns.filter(pattern => {
    if ((0, _utils.isPatternFiltered)(pattern, patternSourceFilter, patternSyncFilter)) {
      return false;
    }
    if (category.name === _utils.allPatternsCategory.name) {
      return true;
    }
    if (category.name === _utils.myPatternsCategory.name && pattern.type === _utils.INSERTER_PATTERN_TYPES.user) {
      return true;
    }
    if (category.name === 'uncategorized') {
      // The uncategorized category should show all the patterns without any category...
      if (!pattern.categories) {
        return true;
      }

      // ...or with no available category.
      return !pattern.categories.some(catName => availableCategories.some(c => c.name === catName));
    }
    return pattern.categories?.includes(category.name);
  }), [allPatterns, availableCategories, category.name, patternSourceFilter, patternSyncFilter]);
  const pagingProps = (0, _usePatternsPaging.default)(currentCategoryPatterns, category, scrollContainerRef);
  const {
    changePage
  } = pagingProps;

  // Hide block pattern preview on unmount.
  // eslint-disable-next-line react-hooks/exhaustive-deps
  (0, _element.useEffect)(() => () => onHover(null), []);
  const onSetPatternSyncFilter = (0, _element.useCallback)(value => {
    setPatternSyncFilter(value);
    changePage(1);
  }, [setPatternSyncFilter, changePage]);
  const onSetPatternSourceFilter = (0, _element.useCallback)(value => {
    setPatternSourceFilter(value);
    changePage(1);
  }, [setPatternSourceFilter, changePage]);
  return (0, _react.createElement)("div", {
    className: "block-editor-inserter__patterns-category-panel"
  }, (0, _react.createElement)(_components.__experimentalVStack, {
    spacing: 2,
    className: "block-editor-inserter__patterns-category-panel-header"
  }, (0, _react.createElement)(_components.__experimentalHStack, null, (0, _react.createElement)(_components.FlexBlock, null, (0, _react.createElement)(_components.__experimentalHeading, {
    level: 4,
    as: "div"
  }, category.label)), (0, _react.createElement)(_patternsFilter.PatternsFilter, {
    patternSyncFilter: patternSyncFilter,
    patternSourceFilter: patternSourceFilter,
    setPatternSyncFilter: onSetPatternSyncFilter,
    setPatternSourceFilter: onSetPatternSourceFilter,
    scrollContainerRef: scrollContainerRef,
    category: category
  })), !currentCategoryPatterns.length && (0, _react.createElement)(_components.__experimentalText, {
    variant: "muted",
    className: "block-editor-inserter__patterns-category-no-results"
  }, (0, _i18n.__)('No results found'))), currentCategoryPatterns.length > 0 && (0, _react.createElement)(_blockPatternsList.default, {
    ref: scrollContainerRef,
    shownPatterns: pagingProps.categoryPatternsAsyncList,
    blockPatterns: pagingProps.categoryPatterns,
    onClickPattern: onClickPattern,
    onHover: onHover,
    label: category.label,
    orientation: "vertical",
    category: category.name,
    isDraggable: true,
    showTitlesAsTooltip: showTitlesAsTooltip,
    patternFilter: patternSourceFilter,
    pagingProps: pagingProps
  }));
}
