// @wordpress/block-editor/build/components/rich-text/embed-handler-picker.native.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = require("react");
var _element = require("@wordpress/element");
var _components = require("@wordpress/components");
var _i18n = require("@wordpress/i18n");
/**
 * WordPress dependencies
 */

const noop = () => {};
const DEFAULT_PICKER_OPTIONS = [{
  id: 'createEmbed',
  label: (0, _i18n.__)('Create embed'),
  value: 'createEmbed',
  onSelect: noop
}, {
  id: 'createLink',
  label: (0, _i18n.__)('Create link'),
  value: 'createLink',
  onSelect: noop
}];
const EmbedHandlerPicker = (0, _element.forwardRef)(({}, ref) => {
  const pickerRef = (0, _element.useRef)();
  const pickerOptions = (0, _element.useRef)(DEFAULT_PICKER_OPTIONS).current;
  const onPickerSelect = (0, _element.useCallback)(value => {
    const selectedItem = pickerOptions.find(item => item.value === value);
    selectedItem.onSelect();
  }, [pickerOptions]);
  (0, _element.useImperativeHandle)(ref, () => ({
    presentPicker: ({
      createEmbed,
      createLink
    }) => {
      pickerOptions[0].onSelect = createEmbed;
      pickerOptions[1].onSelect = createLink;
      pickerRef.current?.presentPicker();
    }
  }));
  return (0, _react.createElement)(_components.Picker, {
    ref: pickerRef,
    options: pickerOptions,
    onChange: onPickerSelect,
    testID: "embed-handler-picker",
    hideCancelButton: true,
    leftAlign: true
  });
});
var _default = exports.default = (0, _element.memo)(EmbedHandlerPicker);
