// @wordpress/block-editor/build/components/rich-text/format-toolbar-container.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = require("react");
var _i18n = require("@wordpress/i18n");
var _components = require("@wordpress/components");
var _blockControls = _interopRequireDefault(require("../block-controls"));
var _formatToolbar = _interopRequireDefault(require("./format-toolbar"));
var _navigableToolbar = _interopRequireDefault(require("../navigable-toolbar"));
/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

function InlineToolbar({
  popoverAnchor
}) {
  return (0, _react.createElement)(_components.Popover, {
    placement: "top",
    focusOnMount: false,
    anchor: popoverAnchor,
    className: "block-editor-rich-text__inline-format-toolbar",
    __unstableSlotName: "block-toolbar"
  }, (0, _react.createElement)(_navigableToolbar.default, {
    className: "block-editor-rich-text__inline-format-toolbar-group"
    /* translators: accessibility text for the inline format toolbar */,
    "aria-label": (0, _i18n.__)('Format tools')
  }, (0, _react.createElement)(_components.ToolbarGroup, null, (0, _react.createElement)(_formatToolbar.default, null))));
}
const FormatToolbarContainer = ({
  inline,
  editableContentElement
}) => {
  if (inline) {
    return (0, _react.createElement)(InlineToolbar, {
      popoverAnchor: editableContentElement
    });
  }

  // Render regular toolbar.
  return (0, _react.createElement)(_blockControls.default, {
    group: "inline"
  }, (0, _react.createElement)(_formatToolbar.default, null));
};
var _default = exports.default = FormatToolbarContainer;
